--- /dev/null
+++ b/net/minecraft/world/entity/vehicle/minecart/MinecartHopper.java.rej
@@ -1,0 +_,15 @@
+diff a/net/minecraft/world/entity/vehicle/minecart/MinecartHopper.java b/net/minecraft/world/entity/vehicle/minecart/MinecartHopper.java	(rejected hunks)
+@@ -102,6 +102,13 @@ public class MinecartHopper extends AbstractMinecartContainer implements Hopper
+         }
+     }
+ 
++    // Mili start - tick every
++    @Override
++    public void inactiveTick() {
++        this.tick();
++    }
++    // Mili end - tick every
++
+     public boolean suckInItems() {
+         if (HopperBlockEntity.suckInItems(this.level(), this)) {
+             this.immunize(); // Paper
