From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Tue, 26 Aug 2025 22:08:54 +0800
Subject: [PATCH] Transformed Configs


diff --git a/src/main/java/me/earthme/luminol/config/modules/experiment/CommandConfig.java b/src/main/java/me/earthme/luminol/config/modules/experiment/CommandConfig.java
index ad2ba680bedb2336de24146676e70a1f64fd9c5a..505f7fe2ac40451e56967c90df9c52fb35f80fdf 100644
--- a/src/main/java/me/earthme/luminol/config/modules/experiment/CommandConfig.java
+++ b/src/main/java/me/earthme/luminol/config/modules/experiment/CommandConfig.java
@@ -13,6 +13,7 @@ public class CommandConfig implements IConfigModule {
     @ConfigInfo(name = "enable_data_command")
     @HotReloadUnsupported
     public static boolean data = false;
+    @TransformedConfig(name = "command_block_enabled", directory = {"experiment", "command"}, originInstance = "lophine")
     @TransformedConfig(name = "enabled", directory = {"experiment", "force_enable_command_block_command_execution"})
     @ConfigInfo(name = "enable_command_block", comments = """
             Force to enable command blocks.
@@ -20,6 +21,9 @@ public class CommandConfig implements IConfigModule {
             DO NOT ENABLE UNLESS YOU KNOW WHAT YOU ARE DOING!!!
             """)
     public static boolean commandBlock = false;
+    @TransformedConfig(name = "enable-waypoint", directory = {"experiment", "waypoint bar"}, originInstance = "lophine")
+    @TransformedConfig(name = "enable-waypoint", directory = {"experiment", "waypoint_bar"}, originInstance = "lophine")
+    @TransformedConfig(name = "waypoint_command_enabled", directory = {"experiment", "command"}, originInstance = "lophine")
     @ConfigInfo(name = "enable_waypoints_and_waypoint_command", comments = """
             Enable waypoint and waypoint command.
             WARN: Still under testing
diff --git a/src/main/java/me/earthme/luminol/config/modules/misc/PaperPacketLimiterConfig.java b/src/main/java/me/earthme/luminol/config/modules/misc/PaperPacketLimiterConfig.java
index aece37eb6406f8ada56428e68cf198c43f8ae309..d5f99fd672e33c16fc3c3e2031ce2e5a5001ab55 100644
--- a/src/main/java/me/earthme/luminol/config/modules/misc/PaperPacketLimiterConfig.java
+++ b/src/main/java/me/earthme/luminol/config/modules/misc/PaperPacketLimiterConfig.java
@@ -11,6 +11,7 @@ import me.earthme.luminol.enums.EnumConfigCategory;
                 "has negative impacts on security"
 )
 public class PaperPacketLimiterConfig implements IConfigModule {
+    @TransformedConfig(name = "unlimit-packet", directory = {"misc", "network"}, originInstance = "lophine")
     @TransformedConfig(name = "force_disable", directory = {"optimizations", "force_disable_packet_limiter_of_paper"})
     @ConfigInfo(name = "force_disable")
     public static boolean forceDisable = false;
