From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 22 Mar 2026 01:39:17 +0800
Subject: [PATCH] Carpet features


diff --git a/src/main/java/me/earthme/luminol/config/ConfigManager.java b/src/main/java/me/earthme/luminol/config/ConfigManager.java
--- a/src/main/java/me/earthme/luminol/config/ConfigManager.java
+++ b/src/main/java/me/earthme/luminol/config/ConfigManager.java
@@ -25,6 +25,7 @@ public class ConfigManager {
     public static void initConfigs() {
         configfiles.put("luminol", ConfigsInstance.of("luminol", "me.earthme.luminol.config.modules"));
         configfiles.put("lophine", ConfigsInstance.of("lophine", "fun.bm.mili.config.modules")); // add lophine global config
+        configfiles.put("mili_carpet", ConfigsInstance.of("mili_carpet", "fun.bm.mili.carpet.config.modules")); // add mili carpet config
         configfiles.put("lophine_carpet", ConfigsInstance.of(
                 new java.io.File("lophine_config"),
                 "lophine_carpet",
