--- /dev/null
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java.rej
@@ -1,0 +_,28 @@
+diff a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java	(rejected hunks)
+@@ -76,7 +76,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
+     }
+ 
+     // Folia start - region threading
+-    private static void attemptTeleport(Entity source, ServerLevel checkWorld, net.minecraft.world.phys.Vec3 to) {
++    private static void attemptTeleport(Entity source, ServerLevel checkWorld, net.minecraft.world.phys.Vec3 to, Entity pearlEntity) {
+         final boolean onPortalCooldown = source.isOnPortalCooldown();
+         // ignore retired callback, in those cases we do not want to teleport
+         source.getBukkitEntity().taskScheduler.schedule(
+@@ -135,7 +135,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
+                             // damage player
+                             teleported.resetFallDistance();
+                             player.resetCurrentImpulseContext();
+-                            player.hurtServer(player.level(), player.damageSources().enderPearl().eventEntityDamager(player), 5.0F); // CraftBukkit // Paper - fix DamageSource API
++                            player.hurtServer(player.level(), player.damageSources().enderPearl().eventEntityDamager(pearlEntity), 5.0F); // CraftBukkit // Paper - fix DamageSource API // Mili fix - #438: pass pearlEntity instead of player
+                             playSound(teleported.level(), to);
+                         } else {
+                             // reset fall damage so that if the entity was falling they do not instantly die
+@@ -176,7 +176,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
+                 }
+                 Entity entity = this.getOwnerRaw();
+                 if (entity != null) {
+-                    attemptTeleport(entity, (ServerLevel)this.level(), this.position());
++                    attemptTeleport(entity, (ServerLevel)this.level(), this.position(), this); // Mili fix - #438: pass this (pearl entity) as damager
+                 }
+                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT);
+                 return;
