From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Tue, 11 Feb 2025 11:46:44 +0800
Subject: [PATCH] Tick regions api


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 43b40cce37655043aea8f154fa2aa688539777d8..dcd32c1e32fd67ec13bb7fca8af20aa665407594 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2071,4 +2071,11 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return POINTERS_SUPPLIER.view(this);
     }
     // Paper end
+
+    // Luminol start - Tick regions api
+    @Override
+    public me.earthme.luminol.api.ThreadedRegionizer getThreadedRegionizer() {
+        return this.world.regioniser.threadedRegionizerAPI;
+    }
+    // Luminol end
 }
