--- /dev/null
+++ b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java.rej
@@ -1,0 +_,10 @@
+diff a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java	(rejected hunks)
+@@ -68,7 +68,7 @@ public class VaultSharedData {
+         Set<UUID> set = config.playerDetector()
+             .detect(level, config.entitySelector(), pos, deactivationRange, false)
+             .stream()
+-            .filter(uuid -> !serverData.getRewardedPlayers().contains(uuid))
++            .filter(uuid -> fun.bm.mili.config.modules.misc.OldMCConfig.disableVaultBlacklist || !serverData.getRewardedPlayers().contains(uuid)) // Mili - disable vault blacklist
+             .collect(Collectors.toSet());
+         if (!this.connectedPlayers.equals(set)) {
+             this.connectedPlayers = set;
