From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sun, 13 Apr 2025 11:07:04 +0800
Subject: [PATCH] Add config to disable heightmap warning


diff --git a/net/minecraft/world/level/levelgen/Heightmap.java b/net/minecraft/world/level/levelgen/Heightmap.java
index 25c537496ccea6fe27111b21fe2e47b3f6a914a1..78398fbd00034c3341be897d37aeedd3860ba905 100644
--- a/net/minecraft/world/level/levelgen/Heightmap.java
+++ b/net/minecraft/world/level/levelgen/Heightmap.java
@@ -128,7 +128,8 @@ public class Heightmap {
         if (raw.length == data.length) {
             System.arraycopy(data, 0, raw, 0, data.length);
         } else {
-            LOGGER.warn("Ignoring heightmap data for chunk {}, size does not match; expected: {}, got: {}", chunk.getPos(), raw.length, data.length);
+            if (!me.earthme.luminol.config.modules.misc.DisableWarningConfig.disableHeightmapWarning)
+                LOGGER.warn("Ignoring heightmap data for chunk {}, size does not match; expected: {}, got: {}", chunk.getPos(), raw.length, data.length);
             primeHeightmaps(chunk, EnumSet.of(type));
         }
     }
