From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Sun, 12 Jan 2025 10:27:31 +0800
Subject: [PATCH] Add tpsbar with chunkhot, membar and regionbar


diff --git a/ca/spottedleaf/moonrise/paper/util/BaseChunkSystemHooks.java b/ca/spottedleaf/moonrise/paper/util/BaseChunkSystemHooks.java
index 874ffbcc0263ded98462a9fd2529e9a877ab8900..4850428ea33c89732b4dede6d8c731368176a2f9 100644
--- a/ca/spottedleaf/moonrise/paper/util/BaseChunkSystemHooks.java
+++ b/ca/spottedleaf/moonrise/paper/util/BaseChunkSystemHooks.java
@@ -122,6 +122,7 @@ public abstract class BaseChunkSystemHooks implements ca.spottedleaf.moonrise.co
 
     @Override
     public void onChunkNotTicking(final LevelChunk chunk, final ChunkHolder holder) {
+        chunk.getChunkHot().clear(); // KioCG
         chunk.getLevel().getCurrentWorldData().removeTickingChunk(chunk); // Folia - region threading
     }
 
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 5bf7635146337df2e89e13c592a8364dbcb4b3f7..89d60d74929a8ee23bcf94134cc71ee25bc58685 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1737,7 +1737,46 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // Folia end - region threading
         //this.tickCount++; // Folia - region threading
         //this.tickRateManager.tick(); // Folia - region threading
+        // KioCG start - ChunkHot
+        final ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet<net.minecraft.world.level.chunk.LevelChunk> chunks = new ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet<>();
+        if (region != null){
+            for (net.minecraft.world.level.chunk.LevelChunk chunk : region.world.getCurrentWorldData().getTickingChunks()) {
+            /* wait for rewrite - temporarily crash fix
+            for (net.minecraft.server.level.ServerChunkCache.ChunkAndHolder chunkAndHolder : region.world.getCurrentWorldData().getTickingChunks()){
+                final net.minecraft.world.level.chunk.LevelChunk chunk = chunkAndHolder.chunk();
+            */
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(region.world, chunk.locX, chunk.locZ)){
+                    continue;
+                }
+
+                chunks.add(chunk);
+            }
+        }
+        if (region != null && io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick() % 20 == 0){
+            final java.util.Iterator<net.minecraft.world.level.chunk.LevelChunk> chunkIterator = chunks.unsafeIterator();
+            while (chunkIterator.hasNext()){
+                final net.minecraft.world.level.chunk.LevelChunk targetChunk = chunkIterator.next();
+
+                targetChunk.getChunkHot().nextTick();
+                targetChunk.getChunkHot().start();
+            }
+        }
+        //KioCG end
         this.tickChildren(hasTimeLeft, region); // Folia - region threading
+        // KioCG start - ChunkHot
+        if (region != null && io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick() % 20 == 0){
+            final java.util.Iterator<net.minecraft.world.level.chunk.LevelChunk> chunkIterator = chunks.unsafeIterator();
+            while (chunkIterator.hasNext()){
+                final net.minecraft.world.level.chunk.LevelChunk targetChunk = chunkIterator.next();
+
+                if (!targetChunk.getChunkHot().isStarted()){
+                    continue;
+                }
+
+                targetChunk.getChunkHot().stop();
+            }
+        }
+        //KioCG end
         if (false && nanos - this.lastServerStatus >= STATUS_EXPIRE_TIME_NANOS) { // Folia - region threading
             this.lastServerStatus = nanos;
             this.status = this.buildServerStatus();
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 84e768799a740151e12d1e0fdcb6feffcdaa2682..24285a4b8a099c153b6d403551c6ef92e70b066b 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -918,6 +918,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     @Override
     public void stopServer() {
+        me.earthme.luminol.functions.bars.GlobalServerBarManager.cancelAll(); // Luminol - turn off all bars
         this.notificationManager().serverShuttingDown();
         super.stopServer();
         //Util.shutdownExecutors(); // Paper - Improved watchdog support; moved into super
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 3471200d54b938ef82f142efc2fcb096d72ec7b4..b8961e6c74170072d148f9db326a9df822e76219 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1438,6 +1438,8 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         final int timerId = isActive ? entity.getType().tickTimerId : entity.getType().inactiveTickTimerId;
         final ca.spottedleaf.leafprofiler.RegionizedProfiler.Handle profiler = io.papermc.paper.threadedregions.TickRegionScheduler.getProfiler();
         profiler.startTimer(timerId);
+        LevelChunk levelChunk = entity.shouldTickHot() ? this.getChunkIfLoaded(entity.moonrise$getSectionX(),entity.moonrise$getSectionZ()) : null; // KioCG
+        if (levelChunk != null) levelChunk.getChunkHot().startTicking(); try { // KioCG
         try {
         // Folia end - profiler
         if (isActive) { // Paper - EAR 2
@@ -1455,6 +1457,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         } else {entity.inactiveTick();} // Paper - EAR 2
         profilerFiller.pop();
         } finally { profiler.stopTimer(timerId); } // Folia - profiler
+            } finally { if (levelChunk != null) levelChunk.getChunkHot().stopTickingAndCount(); } // KioCG
 
         for (Entity entity1 : entity.getPassengers()) {
             this.tickPassenger(entity, entity1, isActive); // Paper - EAR 2
@@ -1474,6 +1477,8 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
             final int timerId = isActive ? passengerEntity.getType().tickTimerId : passengerEntity.getType().inactiveTickTimerId;
             final ca.spottedleaf.leafprofiler.RegionizedProfiler.Handle profiler = io.papermc.paper.threadedregions.TickRegionScheduler.getProfiler();
             profiler.startTimer(timerId);
+            LevelChunk levelChunk = !(passengerEntity instanceof Player) ? this.getChunkIfLoaded(passengerEntity.blockPosition()) : null; // KioCG
+            if (levelChunk != null) levelChunk.getChunkHot().startTicking(); try { // KioCG
             try {
             // Folia end - profiler
             passengerEntity.setOldPosAndRot();
@@ -1508,6 +1513,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
                 this.tickPassenger(passengerEntity, entity, isActive); // Paper - EAR 2
             }
             } finally { profiler.stopTimer(timerId); } // Folia - profiler
+            } finally { if (levelChunk != null) levelChunk.getChunkHot().stopTickingAndCount(); } // KioCG
         }
     }
 
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 01e067a0045e59094bf271c448f158f932199aa3..0eda79be7f2f184f5a769e1200332961c90a822c 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -433,7 +433,9 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     public boolean isRealPlayer; // Paper
     public com.destroystokyo.paper.event.entity.@Nullable PlayerNaturallySpawnCreaturesEvent playerNaturallySpawnedEvent; // Paper - PlayerNaturallySpawnCreaturesEvent
     public org.bukkit.event.player.PlayerQuitEvent.@Nullable QuitReason quitReason = null; // Paper - Add API for quit reason; there are a lot of changes to do if we change all methods leading to the event
-
+    public volatile boolean isTpsBarVisible = false; //Luminol - Tps bar
+    public volatile boolean isMemBarVisible = false; //Luminol - Memory bar
+    public volatile boolean isRegionBarVisible = false; //Luminol - Region bar
     // Paper start - rewrite chunk system
     private ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.PlayerChunkLoaderData chunkLoader;
     private final ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.ViewDistanceHolder viewDistanceHolder = new ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.ViewDistanceHolder();
@@ -935,8 +937,35 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.trackEnteredOrExitedLavaOnVehicle();
         this.updatePlayerAttributes();
         this.advancements.flushDirty(this, true);
+
+        // KioCG start - ChunkHot
+        if (this.tickCount % 20 == 0){
+            this.nearbyChunkHot = this.refreshNearbyChunkHot();
+        }
+        // KioCG end
     }
 
+    // KioCG start - ChunkHot
+    private volatile long nearbyChunkHot = 0;
+
+    public long getNearbyChunkHot() { return this.nearbyChunkHot; }
+
+    private long refreshNearbyChunkHot() {
+        long total = 0L;
+        int searchRadius = ((ServerLevel) this.level()).moonrise$getViewDistanceHolder().getViewDistances().tickViewDistance();
+        for (int i = this.moonrise$getSectionX() - searchRadius; i <= this.moonrise$getSectionX() + searchRadius; ++i) {
+            for (int j = this.moonrise$getSectionZ() - searchRadius; j <= this.moonrise$getSectionZ() + searchRadius; ++j) {
+                net.minecraft.world.level.chunk.LevelChunk targetChunk = this.level().getChunkIfLoaded(i, j);
+                if (targetChunk != null) {
+                    total += targetChunk.getChunkHot().getAverage();
+                }
+            }
+        }
+        return total;
+    }
+    // KioCG end
+
+
     private void updatePlayerAttributes() {
         AttributeInstance attribute = this.getAttribute(Attributes.BLOCK_INTERACTION_RANGE);
         if (attribute != null) {
diff --git a/net/minecraft/world/entity/AreaEffectCloud.java b/net/minecraft/world/entity/AreaEffectCloud.java
index b10762fd050ca03c1d31dc57e42739a1f114ca2f..bd662b124da3a3e8f744c7c7343080010d5c4a58 100644
--- a/net/minecraft/world/entity/AreaEffectCloud.java
+++ b/net/minecraft/world/entity/AreaEffectCloud.java
@@ -419,4 +419,11 @@ public class AreaEffectCloud extends Entity implements TraceableEntity {
             return super.applyImplicitComponent(component, value);
         }
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
 }
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index cbc383f51290ecff4b231c290b8b1756b74b4472..723eb2b296287272fe97499695c644c2e64d0c19 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -6269,4 +6269,6 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         return ((ServerLevel) this.level()).isPositionEntityTicking(this.blockPosition());
     }
     // Paper end - Expose entity id counter
+
+    public boolean shouldTickHot() { return this.tickCount > 20 * 10 && this.isAlive(); } // KioCG
 }
diff --git a/net/minecraft/world/entity/LightningBolt.java b/net/minecraft/world/entity/LightningBolt.java
index e81eabf901e755e1254fff98cdd1e609cf07aab8..2e8925f1b2efde6f4bba97e253500e88a2293df6 100644
--- a/net/minecraft/world/entity/LightningBolt.java
+++ b/net/minecraft/world/entity/LightningBolt.java
@@ -279,4 +279,11 @@ public class LightningBolt extends Entity {
     public final boolean hurtServer(ServerLevel level, DamageSource damageSource, float amount) {
         return false;
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
 }
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index a10bff4f894e93a5307a76982adb145965da1030..6560d19b724fe052366380502c9e9187af43e1fa 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1698,4 +1698,11 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     public float chargeSpeedModifier() {
         return 1.0F;
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return super.shouldTickHot() && (!this.removeWhenFarAway(0.0) || this.isPersistenceRequired() || this.requiresCustomPersistence());
+    }
+    // KioCG end
 }
diff --git a/net/minecraft/world/entity/animal/equine/TraderLlama.java b/net/minecraft/world/entity/animal/equine/TraderLlama.java
index 44f85dc4075e9fc000dc89ba01e2039c01989dde..579a982c4cd8d81ceb385b22756dee04727dfc59 100644
--- a/net/minecraft/world/entity/animal/equine/TraderLlama.java
+++ b/net/minecraft/world/entity/animal/equine/TraderLlama.java
@@ -155,4 +155,11 @@ public class TraderLlama extends Llama {
             super.start();
         }
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return super.shouldTickHot() && !this.canDespawn();
+    }
+    // KioCG end
 }
diff --git a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
index 3eb6ee2ffecbdeb6057d81b7936b80e205db7cdc..3c0fded0e6b87c62cd5189c9e1b367643e94527a 100644
--- a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
+++ b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
@@ -277,4 +277,11 @@ public class WanderingTrader extends AbstractVillager implements Consumable.Over
             return !pos.closerToCenterThan(this.trader.position(), distance);
         }
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
 }
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index b39325b3c11b689a2fe7a57f679ec6a38563b257..67c0f4027a1e8b364694b5ed8b71965f2fe9b359 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1423,6 +1423,13 @@ public abstract class Player extends Avatar implements ContainerUser {
         return true;
     }
 
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
+
     public Either<Player.BedSleepingProblem, Unit> startSleepInBed(BlockPos bedPos) {
         // CraftBukkit start
         return this.startSleepInBed(bedPos, false);
diff --git a/net/minecraft/world/entity/projectile/Projectile.java b/net/minecraft/world/entity/projectile/Projectile.java
index da218b6f84f8501eec8086abf2b61d8552741fc9..ce9d3bb182d8f1b90236492f92a0a10eb1d901f1 100644
--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -487,4 +487,11 @@ public abstract class Projectile extends Entity implements TraceableEntity {
     public interface ProjectileFactory<T extends Projectile> {
         T create(ServerLevel level, LivingEntity owner, ItemStack spawnedFrom);
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
 }
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index dfc3d047f25499820f84d858ccc29c90f2fe3fbc..3cd1224e20b15b77681e2de4ff601724c559557e 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -112,6 +112,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
     // Paper start - rewrite chunk system
     private boolean postProcessingDone;
     private ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder chunkAndHolder;
+    private final com.kiocg.ChunkHot chunkHot = new com.kiocg.ChunkHot(); public com.kiocg.ChunkHot getChunkHot() { return this.chunkHot; } // KioCG
 
     @Override
     public final boolean moonrise$isPostProcessingDone() {
@@ -967,6 +968,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
                     try {
                         ProfilerFiller profilerFiller = Profiler.get();
                         profilerFiller.push(this::getType);
+                        LevelChunk.this.chunkHot.startTicking(); // KioCG
                         profiler.startTimer(timerId); try { // Folia - profiler
                         BlockState blockState = LevelChunk.this.getBlockState(blockPos);
                         if (this.blockEntity.getType().isValid(blockState)) {
@@ -981,7 +983,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
                             }
                             // Paper end - Remove the Block Entity if it's invalid
                         }
-                        } finally { profiler.stopTimer(timerId); } // Folia - profiler
+                        } finally { profiler.stopTimer(timerId);   LevelChunk.this.chunkHot.stopTickingAndCount(); } // Folia - profiler // KioCG
 
                         profilerFiller.pop();
                     } catch (Throwable var5) {
diff --git a/net/minecraft/world/level/redstone/NeighborUpdater.java b/net/minecraft/world/level/redstone/NeighborUpdater.java
index 87512434ccd8d46d8db718228c790465c64b2c20..af21821513f11011e7bf3841d2f79cb99e4cd517 100644
--- a/net/minecraft/world/level/redstone/NeighborUpdater.java
+++ b/net/minecraft/world/level/redstone/NeighborUpdater.java
@@ -57,7 +57,10 @@ public interface NeighborUpdater {
                 return;
             }
             // CraftBukkit end
+            net.minecraft.world.level.chunk.LevelChunk levelChunk = level.getChunkIfLoaded(pos); // KioCG
+            if (levelChunk != null) levelChunk.getChunkHot().startTicking(); try { // KioCG
             state.handleNeighborChanged(level, pos, neighborBlock, orientation, movedByPiston);
+            } finally { if (levelChunk != null) levelChunk.getChunkHot().stopTickingAndCount(); } // KioCG
             // Spigot start
         } catch (StackOverflowError ex) {
             level.lastPhysicsProblem = pos.immutable();
