From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sat, 10 May 2025 19:46:16 +0800
Subject: [PATCH] Fix creative player item pick


diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index 7d8748fd4bdd523a7cf37ef86729da4277416f9a..7c18ce37cf6fda3e5a3040e50a197d086a95b917 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -405,7 +405,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
             Item item1 = item.getItem();
             int count = item.getCount();
             // CraftBukkit start - fire PlayerPickupItemEvent
-            int canHold = entity.getInventory().canHold(item);
+            int canHold = entity.hasInfiniteMaterials() ? count : entity.getInventory().canHold(item); // Luminol - Fix creative item picking
             int remaining = count - canHold;
             boolean flyAtPlayer = false; // Paper
 
