From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mrhua269 <mrhua269@gmail.com>
Date: Sun, 15 Feb 2026 22:39:13 +0800
Subject: [PATCH] Do not pathfind out of tickregion & Recompute or drop path
 when target or next node is out of tickregion


diff --git a/net/minecraft/world/entity/ai/navigation/FlyingPathNavigation.java b/net/minecraft/world/entity/ai/navigation/FlyingPathNavigation.java
index c20d10250e2fdd563d0fcfac47de1e276393b529..7ae3cd98f553f722a47fa26b7cb367de4364e455 100644
--- a/net/minecraft/world/entity/ai/navigation/FlyingPathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/FlyingPathNavigation.java
@@ -62,6 +62,17 @@ public class FlyingPathNavigation extends PathNavigation {
 
             if (!this.isDone()) {
                 Vec3 nextEntityPos = this.path.getNextEntityPos(this.mob);
+                // Luminol - Recompute path when path finding out of current tick region
+                if (me.earthme.luminol.config.modules.fixes.PathfindingFixesConfig.breakDownPathfindingWhenOutOfRegion) {
+                    // we assume that:
+                    // 1. The code above doesn't touch the 'main thread context' with the position from 'this.path'
+                    // 2. The pathfinder could correctly recompute or discard the incorrect target position and this situation is happening rarely
+                    if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.mob.level(), nextEntityPos)) {
+                        this.hasDelayedRecomputation = true;
+                        return;
+                    }
+                }
+                // Luminol end
                 this.mob.getMoveControl().setWantedPosition(nextEntityPos.x, nextEntityPos.y, nextEntityPos.z, this.speedModifier);
             }
         }
diff --git a/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index 0a77b906a426732a2f82e8727ef99b0ebf692238..f0c4d4638128c6e56fd52c3514f1bede701142bb 100644
--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -1,6 +1,8 @@
 package net.minecraft.world.entity.ai.navigation;
 
 import com.google.common.collect.ImmutableSet;
+
+import java.util.HashSet;
 import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
@@ -101,6 +103,19 @@ public abstract class PathNavigation {
     public void recomputePath() {
         if (this.tick - this.timeLastRecompute > 20L) { // Folia - region threading
             if (this.targetPos != null) {
+                // Luminol - Recompute path when path finding out of current tick region
+                if (me.earthme.luminol.config.modules.fixes.PathfindingFixesConfig.breakDownPathfindingWhenOutOfRegion) {
+                    // The target seems to be pointed to a position out of current region, so reset if it is
+                    // also it will interrupt the pathfinding
+                    if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.mob.level(), this.targetPos)) {
+                        // also we need to reset the path field so all of it will be stopped
+                        this.path = null;
+
+                        this.targetPos = null;
+                        return;
+                    }
+                }
+                // Luminol end
                 this.path = null;
                 this.path = this.createPath(this.targetPos, this.reachRange);
                 this.timeLastRecompute = this.tick; // Folia - region threading
@@ -181,6 +196,18 @@ public abstract class PathNavigation {
                 }
             }
             // Paper end - EntityPathfindEvent
+            // Luminol - Do not path find for targets out of current region
+            if (me.earthme.luminol.config.modules.fixes.PathfindingFixesConfig.doNotPathfindToNotOwnedTargets) {
+                // filter the targets not owned by current region
+                targets = new HashSet<>(targets); // well no idea about how to determine if this should be copied to a modifiable one
+                targets.removeIf(pos -> !ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.mob.level(), pos));
+
+                // return if no available (observe the logic in the first if block)
+                if (targets.isEmpty()) {
+                    return null;
+                }
+            }
+            // Luminol end
             ProfilerFiller profilerFiller = Profiler.get();
             profilerFiller.push("pathfind");
             BlockPos blockPos = offsetUpward ? this.mob.blockPosition().above() : this.mob.blockPosition();
@@ -283,6 +310,17 @@ public abstract class PathNavigation {
 
             if (!this.isDone()) {
                 Vec3 tempMobPos = this.path.getNextEntityPos(this.mob);
+                // Luminol - Recompute path when path finding out of current tick region
+                if (me.earthme.luminol.config.modules.fixes.PathfindingFixesConfig.breakDownPathfindingWhenOutOfRegion) {
+                    // we assume that:
+                    // 1. The code above doesn't touch the 'main thread context' with the position from 'this.path'
+                    // 2. The pathfinder could correctly recompute or discard the incorrect target position and this situation is happening rarely
+                    if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.mob.level(), tempMobPos)) {
+                        this.hasDelayedRecomputation = true;
+                        return;
+                    }
+                }
+                // Luminol end
                 this.mob.getMoveControl().setWantedPosition(tempMobPos.x, this.getGroundY(tempMobPos), tempMobPos.z, this.speedModifier);
             }
         }
