From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Sat, 28 Feb 2026 22:05:40 +0800
Subject: [PATCH] Do not load too far poi chunk in poi compete scan

Beeeeeeeeeeeeeeee

diff --git a/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java b/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
index a7552764bec1c9e5a69117fd0be604d1b9cfccfe..938042a643cddc08177e9fe1ea041df74159cd84 100644
--- a/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
+++ b/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
@@ -24,8 +24,18 @@ public class PoiCompetitorScan {
                             return true;
                         }
                         // Folia end - region threading
-                        level.getPoiManager()
-                            .getType(globalPos.pos())
+                        // Luminol start - POI fixes
+                        var blockPosOfJobSite = globalPos.pos();
+                        var sectionPosOfJobSite = net.minecraft.core.SectionPos.asLong(blockPosOfJobSite);
+                        var poiManager = level.getPoiManager();
+                        // we don't care if we should clear the memory of JOB_SITE
+                        // as it will be automatically removed in SetWalkTargetFromBlockMemory
+                        // so simply break down if it's not loaded
+                        var poiChunk = me.earthme.luminol.config.modules.fixes.POIRangeFixes.doNotCompetePOIIfUnloaded ? poiManager.get(sectionPosOfJobSite) : poiManager.getOrLoad(sectionPosOfJobSite);
+                        poiChunk.flatMap(poiSection -> poiSection.getType(blockPosOfJobSite))
+                        // Luminol end - POI fixes
+                        /*level.getPoiManager() // Luminol - POI fixes
+                            .getType(globalPos.pos())*/ // Luminol - POI fixes
                             .ifPresent(
                                 // Paper start - Improve performance of PoiCompetitorScan by unrolling stream
                                 // The previous logic used Stream#reduce to simulate a form of single-iteration bubble sort
