--- /dev/null
+++ b/net/minecraft/world/entity/npc/InventoryCarrier.java.rej
@@ -1,0 +_,20 @@
+diff a/net/minecraft/world/entity/npc/InventoryCarrier.java b/net/minecraft/world/entity/npc/InventoryCarrier.java	(rejected hunks)
+@@ -31,7 +31,17 @@ public interface InventoryCarrier {
+ 
+             mob.onItemPickup(itemEntity);
+             int count = item.getCount();
+-            ItemStack itemStack = inventory.addItem(item);
++            // Mili start - item over-stack util
++            int restoreCount = 0;
++            if (org.leavesmc.leaves.util.ItemOverstackUtils.isStackable(item)) {
++                int vanillaMax = item.getMaxStackSize();
++                restoreCount = Math.max(0, count - vanillaMax);
++                item.setCount(count - restoreCount);
++            }
++            ItemStack itemStack = item.copy();
++            itemStack.setCount(inventory.addItem(item).getCount());
++            itemStack.setCount(itemStack.getCount() + restoreCount);
++            // Mili end - item over-stack util
+             mob.take(itemEntity, count - itemStack.getCount());
+             if (itemStack.isEmpty()) {
+                 itemEntity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
