--- /dev/null
+++ b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java.rej
@@ -1,0 +_,10 @@
+diff a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java	(rejected hunks)
+@@ -134,7 +134,7 @@ public abstract class BaseContainerBlockEntity extends BlockEntity implements Co
+     @Override
+     public void setItem(int slot, ItemStack stack) {
+         this.getItems().set(slot, stack);
+-        stack.limitSize(this.getMaxStackSize(stack));
++        stack.limitSize(this.getMaxStackLeaves(stack)); // Mili - item over-stack util
+         this.setChanged();
+     }
+ 
