--- /dev/null
+++ b/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java.rej
@@ -1,0 +_,29 @@
+diff a/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java b/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java	(rejected hunks)
+@@ -205,6 +210,11 @@ public class BeehiveBlockEntity extends BlockEntity {
+                         1.0F
+                     );
+                 this.level.gameEvent(GameEvent.BLOCK_CHANGE, blockPos, GameEvent.Context.of(bee, this.getBlockState()));
++                // Leaves start - pca
++                if (fun.bm.mili.config.modules.function.protocol.PcaSyncProtocolConfig.enable) {
++                    org.leavesmc.leaves.protocol.PcaSyncProtocol.syncBlockEntityToClient(this);
++                }
++                // Leaves end - pca
+             }
+ 
+             bee.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.ENTER_BLOCK); // CraftBukkit - add Bukkit remove cause
+@@ -328,9 +338,14 @@ public class BeehiveBlockEntity extends BlockEntity {
+                 if (releaseOccupant(level, pos, state, beeData.toOccupant(), null, beeReleaseStatus, savedFlowerPos)) {
+                     flag = true;
+                     iterator.remove();
++                    // Leaves start - pca
++                    if (fun.bm.mili.config.modules.function.protocol.PcaSyncProtocolConfig.enable) {
++                        org.leavesmc.leaves.protocol.PcaSyncProtocol.syncBlockEntityToClient(level.getBlockEntity(pos));
++                    }
++                    // Leaves end - pca
+                 }
+                 // Paper start - Fix bees aging inside; use exitTickCounter to keep actual bee life
+-                else if (level.paperConfig().entities.behavior.cooldownFailedBeehiveReleases) {
++                else if (!fun.bm.mili.config.modules.function.TechnicalSurvivalModeConfig.enabled && level.paperConfig().entities.behavior.cooldownFailedBeehiveReleases) { // Mili - mc technical survival mode
+                     beeData.exitTickCounter = beeData.occupant.minTicksInHive / 2;
+                 }
+                 // Paper end - Fix bees aging inside; use exitTickCounter to keep actual bee life
