--- /dev/null
+++ b/net/minecraft/world/level/block/entity/ChestBlockEntity.java.rej
@@ -1,0 +_,16 @@
+diff a/net/minecraft/world/level/block/entity/ChestBlockEntity.java b/net/minecraft/world/level/block/entity/ChestBlockEntity.java	(rejected hunks)
+@@ -280,4 +280,14 @@ public class ChestBlockEntity extends RandomizableContainerBlockEntity implement
+         items = inventory;
+     }
+     // Leaves end - Lithium Sleeping Block Entity
++
++    // Leaves start - pca
++    @Override
++    public void setChanged() {
++        super.setChanged();
++        if (fun.bm.mili.config.modules.function.protocol.PcaSyncProtocolConfig.enable) {
++            org.leavesmc.leaves.protocol.PcaSyncProtocol.syncBlockEntityToClient(this);
++        }
++    }
++    // Leaves end - pca
+ }
