--- /dev/null
+++ b/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java.rej
@@ -1,0 +_,19 @@
+diff a/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java b/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java	(rejected hunks)
+@@ -113,7 +113,7 @@ public class SculkSensorBlockEntity extends BlockEntity implements GameEventList
+         public boolean canReceiveVibration(ServerLevel level, BlockPos pos, Holder<GameEvent> gameEvent, GameEvent.@Nullable Context context) {
+             return (!pos.equals(this.blockPos) || !gameEvent.is(GameEvent.BLOCK_DESTROY) && !gameEvent.is(GameEvent.BLOCK_PLACE))
+                 && VibrationSystem.getGameEventFrequency(gameEvent) != 0
+-                && SculkSensorBlock.canActivate(SculkSensorBlockEntity.this.getBlockState());
++                && SculkSensorBlock.canActivate(SculkSensorBlockEntity.this.getBlockState(), pos); // Mili - Sound update suppression
+         }
+ 
+         @Override
+@@ -121,7 +121,7 @@ public class SculkSensorBlockEntity extends BlockEntity implements GameEventList
+             ServerLevel level, BlockPos pos, Holder<GameEvent> gameEvent, @Nullable Entity entity, @Nullable Entity playerEntity, float distance
+         ) {
+             BlockState blockState = SculkSensorBlockEntity.this.getBlockState();
+-            if (SculkSensorBlock.canActivate(blockState)) {
++            if (SculkSensorBlock.canActivate(blockState, pos)) { // Mili - Sound update suppression
+                 int gameEventFrequency = VibrationSystem.getGameEventFrequency(gameEvent);
+                 SculkSensorBlockEntity.this.setLastVibrationFrequency(gameEventFrequency);
+                 int redstoneStrengthForDistance = VibrationSystem.getRedstoneStrengthForDistance(distance, this.getListenerRadius());
