--- /dev/null
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java.rej
@@ -1,0 +_,10 @@
+diff a/net/minecraft/world/entity/monster/zombie/Zombie.java b/net/minecraft/world/entity/monster/zombie/Zombie.java	(rejected hunks)
+@@ -314,7 +314,7 @@ public class Zombie extends Monster {
+                 int floor = Mth.floor(this.getX());
+                 int floor1 = Mth.floor(this.getY());
+                 int floor2 = Mth.floor(this.getZ());
+-                EntityType<? extends Zombie> type = this.getType();
++                EntityType<? extends Zombie> type = fun.bm.mili.config.modules.function.OldFeatureConfig.oldZombieReinforcement ? EntityType.ZOMBIE : this.getType(); // Leaves end - old zombie reinforcement
+                 Zombie zombie = type.create(level, EntitySpawnReason.REINFORCEMENT);
+                 if (zombie == null) {
+                     return true;
