--- /dev/null
+++ b/net/minecraft/commands/arguments/item/ItemInput.java.rej
@@ -1,0 +_,17 @@
+diff a/net/minecraft/commands/arguments/item/ItemInput.java b/net/minecraft/commands/arguments/item/ItemInput.java	(rejected hunks)
+@@ -39,11 +39,13 @@ public class ItemInput {
+     public ItemStack createItemStack(int count, boolean allowOversizedStacks) throws CommandSyntaxException {
+         ItemStack itemStack = new ItemStack(this.item, count);
+         itemStack.applyComponents(this.components);
+-        if (allowOversizedStacks && count > itemStack.getMaxStackSize()) {
+-            throw ERROR_STACK_TOO_BIG.create(this.getItemName(), itemStack.getMaxStackSize());
++        // Mili start - item over-stack util
++        if (allowOversizedStacks && count > org.leavesmc.leaves.util.ItemOverstackUtils.getItemStackMaxCount(itemStack)) {
++            throw ERROR_STACK_TOO_BIG.create(this.getItemName(), org.leavesmc.leaves.util.ItemOverstackUtils.getItemStackMaxCount(itemStack));
+         } else {
+             return itemStack;
+         }
++        // Mili end - item over-stack util
+     }
+ 
+     public String serialize(HolderLookup.Provider levelRegistry) {
