--- /dev/null
+++ b/net/minecraft/util/RandomSource.java.rej
@@ -1,0 +_,29 @@
+diff a/net/minecraft/util/RandomSource.java b/net/minecraft/util/RandomSource.java	(rejected hunks)
+@@ -52,13 +52,25 @@ public interface RandomSource {
+ 
+     double nextGaussian();
+ 
++    // Mili start - flattenTriangularDistribution
+     default double triangle(double center, double maxDeviation) {
+-        return center + maxDeviation * (this.nextDouble() - this.nextDouble());
++        if (fun.bm.mili.config.modules.function.OldFeatureConfig.flattenTriangularDistribution) {
++            this.nextDouble();
++            return center + maxDeviation * (-1 + this.nextDouble() * 2);
++        } else {
++            return center + maxDeviation * (this.nextDouble() - this.nextDouble());
++        }
+     }
+ 
+     default float triangle(float center, float maxDeviation) {
+-        return center + maxDeviation * (this.nextFloat() - this.nextFloat());
++        if (fun.bm.mili.config.modules.function.OldFeatureConfig.flattenTriangularDistribution) {
++            this.nextFloat();
++            return center + maxDeviation * (-1 + this.nextFloat() * 2);
++        } else {
++            return center + maxDeviation * (this.nextFloat() - this.nextFloat());
++        }
+     }
++    // Mili end - flattenTriangularDistribution
+ 
+     default void consumeCount(int count) {
+         for (int i = 0; i < count; i++) {
