--- /dev/null
+++ b/net/minecraft/server/commands/GiveCommand.java.rej
@@ -1,0 +_,10 @@
+diff a/net/minecraft/server/commands/GiveCommand.java b/net/minecraft/server/commands/GiveCommand.java	(rejected hunks)
+@@ -54,7 +54,7 @@ public class GiveCommand {
+ 
+     private static int giveItem(CommandSourceStack source, ItemInput item, Collection<ServerPlayer> targets, int count) throws CommandSyntaxException {
+         ItemStack itemStack = item.createItemStack(1, false);
+-        int maxStackSize = itemStack.getMaxStackSize();
++        int maxStackSize = org.leavesmc.leaves.util.ItemOverstackUtils.getItemStackMaxCount(itemStack); // Mili - item over-stack util
+         int i = maxStackSize * 100;
+         if (count > i) {
+             source.sendFailure(Component.translatable("commands.give.failed.toomanyitems", i, itemStack.getDisplayName()));
