--- /dev/null
+++ b/net/minecraft/world/entity/PortalProcessor.java.rej
@@ -1,0 +_,10 @@
+diff a/net/minecraft/world/entity/PortalProcessor.java b/net/minecraft/world/entity/PortalProcessor.java	(rejected hunks)
+@@ -7,7 +7,7 @@ import net.minecraft.world.level.portal.TeleportTransition;
+ import org.jspecify.annotations.Nullable;
+ 
+ public class PortalProcessor {
+-    private final Portal portal;
++    public final Portal portal; // Leaves - private -> public
+     private BlockPos entryPosition;
+     private int portalTime;
+     private boolean insidePortalThisTick;
