From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sat, 24 May 2025 04:27:17 +0800
Subject: [PATCH] Add BlockPos transform to Position support


diff --git a/src/main/java/io/papermc/paper/util/MCUtil.java b/src/main/java/io/papermc/paper/util/MCUtil.java
index 6b3bfe5f1b7e02e5f91406e7bbc53974fe801af3..9e408395885bf063807040110ac46e186d330d70 100644
--- a/src/main/java/io/papermc/paper/util/MCUtil.java
+++ b/src/main/java/io/papermc/paper/util/MCUtil.java
@@ -168,6 +168,12 @@ public final class MCUtil {
         return CraftLocation.toBukkit(pos, world);
     }
 
+    // Luminol start - add BlockPos support
+    public static Location toLocation(Level world, BlockPos pos) {
+        return new Location(world.getWorld(), pos.getX(), pos.getY(), pos.getZ());
+    }
+    // Luminol end - add BlockPos support
+
     /**
      * Converts a NMS World/Vector to Bukkit Location
      *
