--- a/luminol-api/build.gradle.kts
+++ b/luminol-api/build.gradle.kts
@@ -1,8 +_,14 @@
 import paper.libs.com.google.gson.Gson
 
 plugins {
     `java-library`
     `maven-publish`
     idea
+    kotlin("jvm") version "2.3.21"
 }
+
+kotlin {
+    jvmToolchain(21)
+}
+
@@ -104,20 +_,30 @@
             srcDir(generatedDir)
             srcDir(file("../paper-api/src/main/java"))
             srcDir(file("../folia-api/src/main/java"))
+            srcDir(file("../luminol-api/src/main/java"))
         }
         resources {
             srcDir(file("../paper-api/src/main/resources"))
             srcDir(file("../folia-api/src/main/resources"))
+            srcDir(file("../luminol-api/src/main/resources"))
         }
+        kotlin {
+            srcDir("src/main/kotlin")
+        }
     }
     test {
         java {
             srcDir(file("../paper-api/src/test/java"))
             srcDir(file("../folia-api/src/test/java"))
+            srcDir(file("../luminol-api/src/test/java"))
         }
         resources {
             srcDir(file("../paper-api/src/test/resources"))
             srcDir(file("../folia-api/src/test/resources"))
+            srcDir(file("../luminol-api/src/test/resources"))
         }
+        kotlin {
+            srcDir("src/main/kotlin")
+        }
     }
 }
