--- /dev/null
+++ b/net/minecraft/world/entity/monster/Phantom.java.rej
@@ -1,0 +_,22 @@
+diff a/net/minecraft/world/entity/monster/Phantom.java b/net/minecraft/world/entity/monster/Phantom.java	(rejected hunks)
+@@ -238,6 +238,20 @@ public class Phantom extends Mob implements Enemy {
+         return targetingConditions.test(level, this, entity);
+     }
+ 
++    // Mili start - renewable elytra
++    @Override
++    protected void dropFromLootTable(ServerLevel level, DamageSource source, boolean causedByPlayer) {
++        super.dropFromLootTable(level, source, causedByPlayer);
++        if (fun.bm.mili.config.modules.function.OldFeatureConfig.renewableElytra > 0.0D) {
++            if (source.getEntity() instanceof Shulker && this.random.nextDouble() < fun.bm.mili.config.modules.function.OldFeatureConfig.renewableElytra) {
++                net.minecraft.world.item.ItemStack item = new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.ELYTRA);
++                item.setDamageValue(431);
++                this.spawnAtLocation(level, item);
++            }
++        }
++    }
++    // Mili end - renewable elytra
++
+     static enum AttackPhase {
+         CIRCLE,
+         SWOOP;
