--- /dev/null
+++ b/net/minecraft/world/entity/monster/spider/Spider.java.rej
@@ -1,0 +_,20 @@
+diff a/net/minecraft/world/entity/monster/spider/Spider.java b/net/minecraft/world/entity/monster/spider/Spider.java	(rejected hunks)
+@@ -141,6 +141,18 @@ public class Spider extends Monster {
+         this.entityData.set(DATA_FLAGS_ID, b);
+     }
+ 
++    // Mili start - spider jockeys drop gapples
++    @Override
++    protected void dropFromLootTable(net.minecraft.server.level.ServerLevel level, DamageSource damageSource, boolean causedByPlayer) {
++        super.dropFromLootTable(level, damageSource, causedByPlayer);
++        if (fun.bm.mili.config.modules.function.VillagerTradeConfig.spiderJockeysDropGapples > 0.0D) {
++            if (this.hasPassenger(entity -> entity.getType() == EntityType.SKELETON) && this.random.nextDouble() < fun.bm.mili.config.modules.function.VillagerTradeConfig.spiderJockeysDropGapples) {
++                this.spawnAtLocation(level, new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.ENCHANTED_GOLDEN_APPLE));
++            }
++        }
++    }
++    // Mili end - spider jockeys drop gapples
++
+     @Override
+     public @Nullable SpawnGroupData finalizeSpawn(
+         ServerLevelAccessor level, DifficultyInstance difficulty, EntitySpawnReason spawnReason, @Nullable SpawnGroupData spawnGroupData
