From: Mili Bot <mili@example.com>
Date: Thu, 10 Jul 2026 12:00:00 +0800
Subject: [PATCH] Rebrand to Mili

--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -30,6 +30,7 @@ public interface ServerBuildInfo {
     Key BRAND_LUMINOL_ID = Key.key("luminolmc", "luminol");
     // Luminol end
     Key BRAND_LOPHINE_ID = Key.key("luminolmc", "lophine"); // Lophine
+    Key BRAND_MILI_ID = Key.key("xucy10", "mili"); // Mili
 