From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sun, 8 Jun 2025 01:23:55 +0800
Subject: [PATCH] Rebrand to Lophine


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index db5e2f10278c20286779362cfdfadb2f42c724e0..c18c0cb45a193883aa12b5d1abe690e95d289f82 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -29,6 +29,7 @@ public interface ServerBuildInfo {
      */
     Key BRAND_LUMINOL_ID = Key.key("luminolmc", "luminol");
     // Luminol end
+    Key BRAND_LOPHINE_ID = Key.key("luminolmc", "lophine"); // Lophine
 
     /**
      * Gets the {@code ServerBuildInfo}.
