From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mili <dev@bm.fun>
Date: Sun, 8 Jun 2025 01:23:55 +0800
Subject: [PATCH] Rebrand to Mili


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 652ff54e7c50412503725d628bfe72ed03059790..6201dd54d55385609d0f6d8e29536285dc2e65fe 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -17,6 +17,11 @@ public interface ServerBuildInfo {
     /**
      * The brand id for Paper.
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");

+    /**
+     * The brand id for Mili.
+     */
+    Key BRAND_MILI_ID = Key.key("luminolmc", "mili");
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
