--- /dev/null
+++ b/net/minecraft/world/item/FireworkRocketItem.java.rej
@@ -1,0 +_,10 @@
+diff a/net/minecraft/world/item/FireworkRocketItem.java b/net/minecraft/world/item/FireworkRocketItem.java	(rejected hunks)
+@@ -70,7 +70,7 @@ public class FireworkRocketItem extends Item implements ProjectileItem {
+                 com.destroystokyo.paper.event.player.PlayerElytraBoostEvent event = new com.destroystokyo.paper.event.player.PlayerElytraBoostEvent((org.bukkit.entity.Player) player.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemInHand), (org.bukkit.entity.Firework) delayed.projectile().getBukkitEntity(), org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(hand));
+                 if (event.callEvent() && delayed.attemptSpawn()) {
+                     player.awardStat(Stats.ITEM_USED.get(this)); // Moved up from below
+-                    if (player.dropAllLeashConnections(null)) {
++                    if (!fun.bm.mili.config.modules.function.OldFeatureConfig.keepLeashConnectWhenUseFirework && player.dropAllLeashConnections(null)) { // Leaves - Keep leash connection when using firework
+                         level.playSound(null, player, SoundEvents.LEAD_BREAK, SoundSource.NEUTRAL, 1.0F, 1.0F);
+                     }
+                     if (event.shouldConsume() && !player.hasInfiniteMaterials()) {
