--- /dev/null
+++ b/net/minecraft/world/item/enchantment/Enchantment.java.rej
@@ -1,0 +_,24 @@
+diff a/net/minecraft/world/item/enchantment/Enchantment.java b/net/minecraft/world/item/enchantment/Enchantment.java	(rejected hunks)
+@@ -317,7 +317,21 @@ public record Enchantment(Component description, Enchantment.EnchantmentDefiniti
+         applyEffects(
+             this.getEffects(EnchantmentEffectComponents.POST_PIERCING_ATTACK),
+             entityContext(level, enchantmentLevel, entity, entity.position()),
+-            enchantmentEntityEffect -> enchantmentEntityEffect.apply(level, enchantmentLevel, item, entity, entity.position())
++            // Leaves start - EntityLungeEvent
++            enchantmentEntityEffect -> {
++                int lungeLevel = enchantmentLevel;
++                if (entity instanceof LivingEntity livingEntity) {
++                    final io.papermc.paper.event.entity.EntityLungeEvent event = new io.papermc.paper.event.entity.EntityLungeEvent(livingEntity.getBukkitLivingEntity(), lungeLevel);
++                    if (!event.callEvent()) {
++                        return;
++                    }
++
++                    lungeLevel = event.getLungePower();
++                }
++
++                enchantmentEntityEffect.apply(level, lungeLevel, item, entity, entity.position());
++            }
++            // Leaves end - EntityLungeEvent
+         );
+     }
+ 
