From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Mon, 22 Dec 2025 15:40:35 +0800
Subject: [PATCH] Add config for cpu affinity


diff --git a/io/papermc/paper/threadedregions/TickRegionScheduler.java b/io/papermc/paper/threadedregions/TickRegionScheduler.java
index 5a72fa0d72bfdb927293abe921c57f6ae964a6ec..dc8f23697c28573c20f6cc292f7c9d15ae9ad913 100644
--- a/io/papermc/paper/threadedregions/TickRegionScheduler.java
+++ b/io/papermc/paper/threadedregions/TickRegionScheduler.java
@@ -66,7 +66,26 @@ public final class TickRegionScheduler {
 
             @Override
             public Thread newThread(final Runnable run) {
-                final Thread ret = new TickThreadRunner(this.threadGroup, run, "Folia Region Scheduler Thread #" + this.idGenerator.getAndIncrement());
+                // Luminol start - cpu affinity
+                final Runnable actualRun;
+                if (me.earthme.luminol.config.modules.optimizations.CpuAffinityConfig.cpuAffinityEnabled) {
+                    actualRun = new Runnable() {
+                        private boolean affinitySet = false;
+
+                        @Override
+                        public void run() {
+                            if (!this.affinitySet) {
+                                this.affinitySet = true;
+                                net.openhft.affinity.Affinity.setAffinity(me.earthme.luminol.config.modules.optimizations.CpuAffinityConfig.tickRegionAffinityBitSet);
+                            }
+                            run.run();
+                        }
+                    };
+                } else {
+                    actualRun = run;
+                }
+                // Luminol end - cpu affinity
+                final Thread ret = new TickThreadRunner(this.threadGroup, actualRun, "Folia Region Scheduler Thread #" + this.idGenerator.getAndIncrement());
                 ret.setUncaughtExceptionHandler(TickRegionScheduler.this::uncaughtException);
                 return ret;
             }
