From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Tue, 11 Aug 2026 18:56:13 +0800
Subject: [PATCH] Add config for entity to clean memory value which is not
 belong to current tickregion


diff --git a/net/minecraft/world/entity/ai/Brain.java b/net/minecraft/world/entity/ai/Brain.java
index e097bc1268b4050e3948413c52dea69950858458..2a3ae7bc52880ba089f1632f3b3de877666d4a09 100644
--- a/net/minecraft/world/entity/ai/Brain.java
+++ b/net/minecraft/world/entity/ai/Brain.java
@@ -383,7 +383,7 @@ public class Brain<E extends LivingEntity> {
     }
 
     public void tick(final ServerLevel level, final E body) {
-        this.forgetOutdatedMemories();
+        this.forgetOutdatedMemories(body); // Luminol - Add config to force clean entity memory that don't belong to current tick region
         this.tickSensors(level, body);
         this.startEachNonRunningBehavior(level, body);
         this.tickEachRunningBehavior(level, body);
@@ -395,8 +395,33 @@ public class Brain<E extends LivingEntity> {
         }
     }
 
-    private void forgetOutdatedMemories() {
-        this.memories.values().forEach(MemorySlot::tick);
+    private void forgetOutdatedMemories(final E owner) { // Luminol - Add config to force clean entity memory that don't belong to current tick region
+        for (Entry<MemoryModuleType<?>, MemorySlot<?>> entry : this.memories.entrySet()) {
+            MemorySlot<?> slot = entry.getValue();
+            if (slot.hasValue()) {
+                // Luminol start - Add config to force clean entity memory that don't belong to current tick region
+                final Object value = slot.value();
+                final net.minecraft.world.level.Level ownerLevel = owner.level();
+
+                // type: entity
+                if (me.earthme.luminol.config.modules.fixes.ForceCleanupEntityBrainMemoryConfig.enabledForEntity && value instanceof LivingEntity entity) {
+                    if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                        this.eraseMemory(entry.getKey());
+                        continue;
+                    }
+                }
+
+                // type: block_pos
+                if (me.earthme.luminol.config.modules.fixes.ForceCleanupEntityBrainMemoryConfig.enabledForBlockPos && value instanceof net.minecraft.core.BlockPos blockPos) {
+                    if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(ownerLevel, blockPos)) {
+                        this.eraseMemory(entry.getKey());
+                        continue;
+                    }
+                }
+                // Luminol end
+            }
+            slot.tick();
+        }
     }
 
     public void stopAll(final ServerLevel level, final E body) {
