From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Fri, 16 Jan 2026 02:34:43 +0800
Subject: [PATCH] Add config to enable save-all command


diff --git a/io/papermc/paper/threadedregions/RegionizedServer.java b/io/papermc/paper/threadedregions/RegionizedServer.java
index 17108636a5728efe98a064001e00e3ca098b1f51..7748e9ac04fbd3f5259d92fc908d90889cba86c8 100644
--- a/io/papermc/paper/threadedregions/RegionizedServer.java
+++ b/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -350,6 +350,8 @@ public final class RegionizedServer {
         // tick player ping sample
         this.tickPlayerSample();
 
+        fun.bm.mili.utils.SaveAllUtil.checkTimeout(); // Lophine - save-all fix
+
         // tick worlds
         for (final ServerLevel world : this.worlds) {
             this.globalTick(world, tickCount);
diff --git a/io/papermc/paper/threadedregions/TickRegions.java b/io/papermc/paper/threadedregions/TickRegions.java
index 02c004bc7f9a0203523ce4b27cd766d6bdc478c0..3d86e21198f2888422330b6f7c27712f40a6aabc 100644
--- a/io/papermc/paper/threadedregions/TickRegions.java
+++ b/io/papermc/paper/threadedregions/TickRegions.java
@@ -190,6 +190,8 @@ public final class TickRegions implements ThreadedRegionizer.RegionCallbacks<Tic
         public final ThreadedRegionizer.ThreadedRegion<TickRegionData, TickRegionSectionData> region;
         public final ServerLevel world;
 
+        public long lastSavedTime = 0; // Lophine - save-all fix
+
         // generic regionised data
         private final Reference2ReferenceOpenHashMap<RegionizedData<?>, Object> regionizedData = new Reference2ReferenceOpenHashMap<>();
 
diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index 6cb0d4daf0f275b95d5ca09275e40ab32258eb39..48628174299f8f18170198df1834d6174bd1fafa 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -294,7 +294,11 @@ public class Commands {
             PardonCommand.register(this.dispatcher);
             PardonIpCommand.register(this.dispatcher);
             //PerfCommand.register(this.dispatcher); // Folia - region threading - TODO later
-            //SaveAllCommand.register(this.dispatcher); // Folia - region threading - TODO later
+            // Lophine start - Add a config to enable save all command
+            if (fun.bm.mili.config.modules.experiment.CommandConfig.saveAll) {
+                SaveAllCommand.register(this.dispatcher); // Folia - region threading - TODO later
+            }
+            // Lophine end - Add a config to enable save all command
             SaveOffCommand.register(this.dispatcher);
             SaveOnCommand.register(this.dispatcher);
             SetPlayerIdleTimeoutCommand.register(this.dispatcher);
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 8ce5cdda24d63b33b39e32f9712fafa3917f27c5..25324d81bbe876dbe47b33bd5749b72e4283bf80 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1656,6 +1656,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     // Folia start - region threading
     public void tickServer(long startTime, long scheduledEnd, long targetBuffer,
                            io.papermc.paper.threadedregions.TickRegions.TickRegionData region) {
+        // Lophine start - save-all fix
+        if (fun.bm.mili.utils.SaveAllUtil.isSaving()) {
+            fun.bm.mili.utils.SaveAllUtil.postRegionSave(region);
+        }
+        // Lophine end - save-all fix
         final ca.spottedleaf.leafprofiler.RegionizedProfiler.Handle foliaProfiler = io.papermc.paper.threadedregions.TickRegionScheduler.getProfiler(); // Folia - profiler
         // Folia end - region threading
         org.spigotmc.WatchdogThread.tick(); // Spigot
diff --git a/net/minecraft/server/commands/SaveAllCommand.java b/net/minecraft/server/commands/SaveAllCommand.java
index e37b71402179c73146ff67babc228ccfb996f8f4..cfdb60bd89d83282f2202a5585bac22a6002e138 100644
--- a/net/minecraft/server/commands/SaveAllCommand.java
+++ b/net/minecraft/server/commands/SaveAllCommand.java
@@ -21,7 +21,9 @@ public class SaveAllCommand {
     }
 
     private static int saveAll(CommandSourceStack source, boolean flush) throws CommandSyntaxException {
-        source.sendSuccess(() -> Component.translatable("commands.save.saving"), false);
+        // Lophine start - save all
+        fun.bm.mili.utils.SaveAllUtil.preSaveAll(source, flush);
+/*        source.sendSuccess(() -> Component.translatable("commands.save.saving"), false);
         MinecraftServer server = source.getServer();
         boolean flag = server.saveEverything(true, flush, true);
         if (!flag) {
@@ -29,6 +31,12 @@ public class SaveAllCommand {
         } else {
             source.sendSuccess(() -> Component.translatable("commands.save.success"), true);
             return 1;
-        }
+        }*/
+        return 1;
     }
+
+    public static SimpleCommandExceptionType getErrorFailed() {
+        return ERROR_FAILED;
+    }
+    // Lophine end - save all
 }
