From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Tue, 11 Aug 2026 19:05:38 +0800
Subject: [PATCH] Add config to disable entity tick catchers


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 978c53a41766362d3de07963c8d98409d1c78bb8..93bc8fe11f920b48479510a639f21b9b9a816ca6 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1573,6 +1573,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         try {
             tick.accept(entity);
         } catch (Throwable t) {
+            if (me.earthme.luminol.config.modules.experiment.DisableEntityCatchConfig.enabled) throw t; // Luminol
             // Paper start - Prevent block entity and entity crashes
             final String msg = String.format("Entity threw exception at %s:%s,%s,%s", io.papermc.paper.util.MCUtil.getLevelName(entity.level()), entity.getX(), entity.getY(), entity.getZ());
             MinecraftServer.LOGGER.error(msg, t);
