From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sun, 18 May 2025 03:48:53 +0800
Subject: [PATCH] Add config to modify tripwire behavior


diff --git a/net/minecraft/world/level/block/TripWireHookBlock.java b/net/minecraft/world/level/block/TripWireHookBlock.java
index 292fe41967db74884e5937cb125f1e022ccfb6bd..42a5e9a7f5d604e4821a116ba13319df052faf97 100644
--- a/net/minecraft/world/level/block/TripWireHookBlock.java
+++ b/net/minecraft/world/level/block/TripWireHookBlock.java
@@ -199,10 +199,17 @@ public class TripWireHookBlock extends Block {
                     BlockPos blockPos1 = pos.relative(direction, i2);
                     BlockState blockState2 = blockStates[i2];
                     if (blockState2 != null) {
-                        BlockState blockState3 = level.getBlockState(blockPos1);
-                        if (blockState3.is(Blocks.TRIPWIRE) || blockState3.is(Blocks.TRIPWIRE_HOOK)) {
-                            if (!io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates || !blockState3.is(Blocks.TRIPWIRE))  level.setBlock(blockPos1, blockState2.trySetValue(ATTACHED, flag2), Block.UPDATE_ALL); // Paper - prevent tripwire from updating
+                        // Luminol start - tripwire and tripwireHook dupe
+                        if (me.earthme.luminol.config.modules.function.TripwireBehaviorConfig.enabled) {
+                            level.setBlock(blockPos1, blockState2.trySetValue(ATTACHED, flag2), 3);
+                            level.getBlockState(blockPos1);
+                        } else {
+                            BlockState blockState3 = level.getBlockState(blockPos1);
+                            if (blockState3.is(Blocks.TRIPWIRE) || blockState3.is(Blocks.TRIPWIRE_HOOK)) {
+                                if (!io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates || !blockState3.is(Blocks.TRIPWIRE))  level.setBlock(blockPos1, blockState2.trySetValue(ATTACHED, flag2), 3); // Paper - prevent tripwire from updating
+                            }
                         }
+                        // Luminol end - tripwire and tripwireHook dupe
                     }
                 }
             }
diff --git a/net/minecraft/world/level/levelgen/feature/EndPlatformFeature.java b/net/minecraft/world/level/levelgen/feature/EndPlatformFeature.java
index 2fa0b230ab631e71645022e881779b9625fdfb0d..20b3bba2135e3abf2b2a4d294875749c007872ad 100644
--- a/net/minecraft/world/level/levelgen/feature/EndPlatformFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/EndPlatformFeature.java
@@ -28,6 +28,10 @@ public class EndPlatformFeature extends Feature<NoneFeatureConfiguration> {
         // CraftBukkit end
         BlockPos.MutableBlockPos mutableBlockPos = pos.mutable();
 
+        // Luminol start - tripwire behavior modifier
+        java.util.List<BlockPos> blockList1 = new java.util.ArrayList<>();
+        java.util.List<BlockPos> blockList2 = new java.util.ArrayList<>();
+        boolean flag21 = me.earthme.luminol.config.modules.function.TripwireBehaviorConfig.behaviorMode == me.earthme.luminol.enums.EnumTripwireBehavior.VANILLA21;
         for (int i = -2; i <= 2; i++) {
             for (int i1 = -2; i1 <= 2; i1++) {
                 for (int i2 = -1; i2 < 3; i2++) {
@@ -35,7 +39,29 @@ public class EndPlatformFeature extends Feature<NoneFeatureConfiguration> {
                     Block block = i2 == -1 ? Blocks.OBSIDIAN : Blocks.AIR;
                     if (!blockList.getBlockState(blockPos).is(block)) { // CraftBukkit
                         if (dropBlocks) {
-                            blockList.destroyBlock(blockPos, true, null); // CraftBukkit
+                            boolean flag = false;
+                            if (me.earthme.luminol.config.modules.function.TripwireBehaviorConfig.enabled) {
+                                switch (me.earthme.luminol.config.modules.function.TripwireBehaviorConfig.behaviorMode) {
+                                    case me.earthme.luminol.enums.EnumTripwireBehavior.VANILLA20: {
+                                        flag = true;
+                                    }
+                                    case me.earthme.luminol.enums.EnumTripwireBehavior.MIXED: {
+                                        net.minecraft.world.level.block.state.BlockState state = level.getBlockState(blockPos);
+                                        if (state.is(Blocks.TRIPWIRE)) {
+                                            if (state.getValue(net.minecraft.world.level.block.TripWireBlock.DISARMED)) {
+                                                flag = true;
+                                                blockList2.add(blockPos.immutable());
+                                            }
+                                            if (!flag) {
+                                                flag = checkString(blockList2, blockPos);
+                                            }
+                                        }
+                                    }
+                                    default: {} // Luminol - 1.21 & default Logic - default empty
+                                }
+                            }
+                            if (flag) blockList1.add(blockPos.immutable());
+                            else blockList.destroyBlock(blockPos, true, null); // CraftBukkit
                         }
 
                         blockList.setBlock(blockPos, block.defaultBlockState(), Block.UPDATE_ALL); // CraftBukkit
@@ -53,11 +79,30 @@ public class EndPlatformFeature extends Feature<NoneFeatureConfiguration> {
             if (portalEvent.isCancelled()) return;
         }
 
-        if (dropBlocks) {
-            blockList.placeBlocks(state -> level.destroyBlock(state.getPosition(), true, null));
+        if (flag21 || !me.earthme.luminol.config.modules.function.TripwireBehaviorConfig.enabled) {
+            if (dropBlocks) {
+                blockList.placeBlocks(state -> level.destroyBlock(state.getPosition(), true, null));
+            } else {
+                blockList.placeBlocks();
+            }
         } else {
+            if (dropBlocks) {
+                blockList.getSnapshotBlocks().forEach((state) -> {
+                                        level.destroyBlock(state.getPosition(), !blockList1.contains(state.getPosition()), null);
+                });
+                // Luminol - prevent tripwire dupe in end platform generate
+            }
             blockList.placeBlocks();
         }
         // CraftBukkit end
     }
+
+    private static boolean checkString(java.util.List<BlockPos> blockList, BlockPos blockPos) {
+        for (BlockPos pos : blockList) {
+            if (pos.getY() != blockPos.getY()) continue;
+            if (pos.getX() == blockPos.getX() || pos.getZ() == blockPos.getZ()) return true;
+        }
+        return false;
+    }
+    // Luminol end - tripwire behavior modifier
 }
