From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xucy10 <someonm201141923@outlook.com>
Date: Tue, 21 Jul 2026 13:48:00 +0800
Subject: [PATCH] Leaves-Prevent-loss-of-item-drops-due-to-update-supp


diff --git a/net/minecraft/server/level/ServerPlayerGameMode.java b/net/minecraft/server/level/ServerPlayerGameMode.java
index 3301e521a79adbc6dd76be05e69cce1c97f8697e..8dc1db0e6bafba19e07a7fa78f3920d572d38466 100644
--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -386,7 +386,18 @@ public class ServerPlayerGameMode {
                 this.level.getCurrentWorldData().captureDrops = new java.util.ArrayList<>(); // Folia - region threading
                 // CraftBukkit end
                 BlockState blockState1 = block.playerWillDestroy(this.level, pos, blockState, this.player);
-                boolean flag = this.level.removeBlock(pos, false);
+                // Leaves start - Prevent loss of item drops due to update suppression when breaking blocks
+                boolean flag;
+                org.leavesmc.leaves.util.UpdateSuppressionException ex = null;
+                try {
+                    flag = this.level.removeBlock(pos, false);
+                } catch (org.leavesmc.leaves.util.UpdateSuppressionException e) {
+                    ex = e;
+                    ex.provideBlock(level, pos, block);
+                    ex.providePlayer(this.player);
+                    flag = false;
+                }
+                // Leaves end - Prevent loss of item drops due to update suppression when breaking blocks
                 if (SharedConstants.DEBUG_BLOCK_BREAK) {
                     LOGGER.info("server broke {} {} -> {}", pos, blockState1, this.level.getBlockState(pos));
                 }
@@ -418,6 +429,7 @@ public class ServerPlayerGameMode {
                 if (event.isDropItems()) {
                     org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDropItemEvent(bblock, state, this.player, itemsToDrop); // Paper - capture all item additions to the world
                 }
+                if (ex != null) throw ex; // Leaves - Prevent loss of item drops due to update suppression when breaking blocks
 
                 // Drop event experience
                 if (flag) {
