From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Tue, 11 Aug 2026 19:11:45 +0800
Subject: [PATCH] Add config gui support


diff --git a/net/minecraft/server/dialog/action/ParsedTemplate.java b/net/minecraft/server/dialog/action/ParsedTemplate.java
index 8a720ef5cd10bf04e97fb34c1ca0f0b265e98468..8e8309a6ee2f19348c0ebefe776a8d7f7cc2a84f 100644
--- a/net/minecraft/server/dialog/action/ParsedTemplate.java
+++ b/net/minecraft/server/dialog/action/ParsedTemplate.java
@@ -13,7 +13,7 @@ public class ParsedTemplate {
     private final String raw;
     private final StringTemplate parsed;
 
-    private ParsedTemplate(final String raw, final StringTemplate parsed) {
+    public ParsedTemplate(final String raw, final StringTemplate parsed) { // Luminol - private -> public
         this.raw = raw;
         this.parsed = parsed;
     }
