From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kercute <A3167717663@hotmail.com>
Date: Sun, 16 Mar 2025 23:31:41 +0800
Subject: [PATCH] Add force the data command to be enabled config


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index c859197d643fd96ea2a5ef51837abc457e27f002..0ae2080496ca149d0bc2ad97ae6107fe819c73f0 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -200,7 +200,9 @@ public class Commands {
         ClearInventoryCommands.register(this.dispatcher, context);
         //CloneCommands.register(this.dispatcher, context); // Folia - region threading - TODO
         DamageCommand.register(this.dispatcher, context);
-        //DataCommands.register(this.dispatcher); // Folia - region threading - TODO
+        if(me.earthme.luminol.config.modules.experiment.CommandConfig.data) { // Luminol - Config for data command
+        DataCommands.register(this.dispatcher); // Folia - region threading - TODO // Luminol - Config for data command
+        } // Luminol - Config for data command
         //DataPackCommand.register(this.dispatcher, context); // Folia - region threading - TODO
         //DebugCommand.register(this.dispatcher); // Folia - region threading - TODO
         DefaultGameModeCommands.register(this.dispatcher);
