From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xucy10 <someonm201141923@outlook.com>
Date: Tue, 21 Jul 2026 13:48:02 +0800
Subject: [PATCH] Add-removed-check-before-all-checks-start-in-addEffe


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index fde90a8ae313319b1b458678ccdb6a2cb6caffd7..56fe7ccdccf21bb58a4da0bc2f22517de85bd291 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1186,6 +1186,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     public boolean addEffect(MobEffectInstance effectInstance, @Nullable Entity entity, EntityPotionEffectEvent.Cause cause, boolean fireEvent) {
         // Paper end - Don't fire sync event during generation
         // org.spigotmc.AsyncCatcher.catchOp("effect add"); // Spigot // Paper - move to API
+        if (this.isRemoved()) return false; // Mili - shouldn't be possible to add effects to removed entity
         if (!this.hasNullCallback()) ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this, "Cannot add effects to entities asynchronously"); // Folia - region threading
         if (this.isTickingEffects) {
             this.effectsToProcess.add(new ProcessableEffect(effectInstance, cause));
