From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sat, 3 Jan 2026 14:35:56 +0800
Subject: [PATCH] Add missing teleportation event APIs for folia


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 5c2930d586ed8f8e31c41e2805dcc189a1d62424..e842358b52366c1500e553c6f79a0ffd243f5d91 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1768,6 +1768,9 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     if (respawnComplete != null) {
                         respawnComplete.accept(ServerPlayer.this);
                     }
+                    // Luminol - Add missing teleportation apis
+                    new me.earthme.luminol.api.entity.player.PostPlayerRespawnEvent(ServerPlayer.this.getBukkitEntity()).callEvent();
+                    // Luminol end
                 }
             );
         });
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 8f6ad64cf1036d521e33414c322f47498cc385c2..ab10f5dcf32a86520bb2e23a9aef9be13d118a6c 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4490,6 +4490,31 @@ public abstract class Entity
         }
 
         // TODO any events that can modify go HERE
+        // Luminol start - Add missing teleportation apis
+        org.bukkit.Location destinationLoc;
+
+        if (pitch == null) {
+            if (yaw == null) {
+                destinationLoc = io.papermc.paper.util.MCUtil.toLocation(destination, pos, Float.NaN, Float.NaN);
+            } else {
+                destinationLoc = io.papermc.paper.util.MCUtil.toLocation(destination, pos, yaw, Float.NaN);
+            }
+        }else {
+            if (yaw == null) {
+                destinationLoc = io.papermc.paper.util.MCUtil.toLocation(destination, pos, Float.NaN, pitch);
+            }else {
+                destinationLoc = io.papermc.paper.util.MCUtil.toLocation(destination, pos, yaw, pitch);
+            }
+        }
+
+        final me.earthme.luminol.api.entity.EntityTeleportAsyncEvent wrapped = new me.earthme.luminol.api.entity.EntityTeleportAsyncEvent(
+                this.getBukkitEntity(),
+                cause,
+                destinationLoc
+        );
+
+        wrapped.callEvent();
+        // Luminol end
 
         // check for same region
         if (destination == this.level()
@@ -4600,6 +4625,17 @@ public abstract class Entity
             case END: {
                 if (destination.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.END) {
                     BlockPos targetPos = ServerLevel.END_SPAWN_POINT;
+                    // Luminol start - Add missing teleportation apis
+                    final org.bukkit.Location orginalPortalLocation = io.papermc.paper.util.MCUtil.toLocation(origin, originPortal);
+                    final org.bukkit.Location targetPortalLocation = io.papermc.paper.util.MCUtil.toLocation(destination, targetPos);
+
+                    final me.earthme.luminol.api.portal.PortalLocateEvent portalLocateEvent = new me.earthme.luminol.api.portal.PortalLocateEvent(
+                            orginalPortalLocation,
+                            targetPortalLocation
+                    );
+
+                    portalLocateEvent.callEvent();
+                    // Luminol end
                     // need to load chunks so we can create the platform
                     destination.moonrise$loadChunksAsync(
                         targetPos, 16, // load 16 blocks to be safe from block physics
@@ -4623,6 +4659,17 @@ public abstract class Entity
                     );
                 } else {
                     BlockPos spawnPos = destination.getRespawnData().pos();
+                    // Luminol start - Add missing teleportation apis
+                    final org.bukkit.Location orginalPortalLocation = io.papermc.paper.util.MCUtil.toLocation(origin, originPortal);
+                    final org.bukkit.Location targetPortalLocation = io.papermc.paper.util.MCUtil.toLocation(destination, spawnPos);
+
+                    final me.earthme.luminol.api.portal.PortalLocateEvent portalLocateEvent = new me.earthme.luminol.api.portal.PortalLocateEvent(
+                            orginalPortalLocation,
+                            targetPortalLocation
+                    );
+
+                    portalLocateEvent.callEvent();
+                    // Luminol end
                     // need to load chunk for heightmap
                     destination.moonrise$loadChunksAsync(
                         spawnPos, 0,
@@ -4676,7 +4723,17 @@ public abstract class Entity
                 WorldBorder destinationBorder = destination.getWorldBorder();
                 double dimensionScale = net.minecraft.world.level.dimension.DimensionType.getTeleportationScale(origin.dimensionType(), destination.dimensionType());
                 BlockPos targetPos = destination.getWorldBorder().clampToBounds(this.getX() * dimensionScale, this.getY(), this.getZ() * dimensionScale);
+                // Luminol start - Add missing teleportation apis
+                final org.bukkit.Location orginalPortalLocation = io.papermc.paper.util.MCUtil.toLocation(origin, originPortal);
+                final org.bukkit.Location targetPortalLocation = io.papermc.paper.util.MCUtil.toLocation(destination, targetPos);
 
+                final me.earthme.luminol.api.portal.PortalLocateEvent portalLocateEvent = new me.earthme.luminol.api.portal.PortalLocateEvent(
+                        orginalPortalLocation,
+                        targetPortalLocation
+                );
+
+                portalLocateEvent.callEvent();
+                // Luminol end
                 ca.spottedleaf.concurrentutil.completable.CallbackCompletable<BlockUtil.FoundRectangle> portalFound
                     = new ca.spottedleaf.concurrentutil.completable.CallbackCompletable<>();
 
@@ -4813,6 +4870,15 @@ public abstract class Entity
         if (!this.canPortalAsync(destination, takePassengers)) {
             return false;
         }
+        // Luminol start - Add missing teleportation events
+        if (!new me.earthme.luminol.api.entity.PreEntityPortalEvent(
+                this.getBukkitEntity(),
+                io.papermc.paper.util.MCUtil.toLocation(this.level, portalPos),
+                destination.getWorld()
+        ).callEvent()) {
+            return false;
+        }
+        // Luminol end
 
         Vec3 initialPosition = this.position();
         ChunkPos initialPositionChunk = new ChunkPos(
diff --git a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index e29e379f9b2238eafe27545681bb19c710003bcc..fdf220424f7693a7ea3ae385ddd771720825b0f7 100644
--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -192,6 +192,18 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
             portalTile.trySearchForExit(portalWorld, portalPos);
             return false;
         }
+        // Luminol start - Add missing teleportation apis
+        final org.bukkit.Location orginalPortalLocation = io.papermc.paper.util.MCUtil.toLocation(toTeleport.level(), portalPos);
+        final org.bukkit.Location targetPortalLocation = io.papermc.paper.util.MCUtil.toLocation(portalWorld, teleportPos);
+
+        final me.earthme.luminol.api.portal.PortalLocateEvent portalLocateEvent = new me.earthme.luminol.api.portal.PortalLocateEvent(
+                orginalPortalLocation,
+                targetPortalLocation
+        );
+
+        portalLocateEvent.callEvent();
+        // Luminol end
+
 
         // note: we handle the position from the TeleportTransition
         net.minecraft.world.level.portal.TeleportTransition teleport = net.minecraft.world.level.block.EndGatewayBlock.getTeleportTransition(
