From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Mon, 14 Jul 2025 18:52:13 +0800
Subject: [PATCH] Async protocol switching optimization


diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 1efe83262ba0cfdc0e529c387661643cb3709052..9f444988d40208a1bfa9f147ef80489f81115e93 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -349,6 +349,129 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
         }
     }
 
+    // Luminol start - async protocol switcher
+    public <T extends PacketListener> void setupInboundProtocolAsync(
+            ProtocolInfo<T> protocol,
+            T packetListener,
+            @Nullable Runnable callback,
+            boolean resumeAutoReading
+    ) {
+        this.validateListener(protocol, packetListener);
+        if (protocol.flow() != this.getReceiving()) {
+            throw new IllegalStateException("Invalid inbound protocol: " + protocol.id());
+        } else {
+            this.packetListener = packetListener;
+            this.disconnectListener = null;
+
+            UnconfiguredPipelineHandler.InboundConfigurationTask configMessage = UnconfiguredPipelineHandler.setupInboundProtocol(protocol);
+            BundlerInfo bundlerInfo = protocol.bundlerInfo();
+            if (bundlerInfo != null) {
+                PacketBundlePacker newBundler = new PacketBundlePacker(bundlerInfo);
+                configMessage = configMessage.andThen(context -> context.pipeline().addAfter("decoder", "bundler", newBundler));
+            }
+
+            // Here we could execute it in event loop async to prevent waiting io task on main
+            // stop reading new packets to prevent some packets came into pipeline too early
+            this.channel.config().setAutoRead(false);
+
+            // do our configuration task
+            final UnconfiguredPipelineHandler.InboundConfigurationTask finalInboundConfigurationTask = configMessage;
+            Runnable toExecute = () -> this.channel.writeAndFlush(finalInboundConfigurationTask).addListener(future -> {
+                try {
+                    if (future.isSuccess()) {
+                        if (callback != null) callback.run(); // retire callback if there have one
+                        return;
+                    }
+
+                    final Throwable ex = future.cause();
+
+                    // here we process our exceptions like that blocking one
+                    if (ex instanceof ClosedChannelException) {
+                        LOGGER.info("Connection closed during protocol change");
+                    } else {
+                        this.channel.pipeline().fireExceptionCaught(ex);
+                    }
+                }finally {
+                    // reset auto back and resume reading if needed
+                    if (resumeAutoReading) {
+                        this.channel.config().setAutoRead(true);
+                        this.channel.read();
+                    }
+                }
+            });
+
+            // we need to do this inside the event loop
+            if (!this.channel.eventLoop().inEventLoop()) {
+                this.channel.eventLoop().execute(toExecute);
+                return;
+            }
+
+            toExecute.run();
+        }
+    }
+
+    public void setupOutboundProtocolAsync(
+            ProtocolInfo<?> protocol,
+            @Nullable Runnable callback,
+            boolean resumeAutoReading
+    ) {
+        if (protocol.flow() != this.getSending()) {
+            throw new IllegalStateException("Invalid outbound protocol: " + protocol.id());
+        } else {
+            UnconfiguredPipelineHandler.OutboundConfigurationTask configMessage = UnconfiguredPipelineHandler.setupOutboundProtocol(protocol);
+            BundlerInfo bundlerInfo = protocol.bundlerInfo();
+            if (bundlerInfo != null) {
+                PacketBundleUnpacker newUnbundler = new PacketBundleUnpacker(bundlerInfo);
+                configMessage = configMessage.andThen(
+                        context -> context.pipeline().addAfter("encoder", "unbundler", newUnbundler)
+                );
+            }
+
+            boolean isLoginProtocol = protocol.id() == ConnectionProtocol.LOGIN;
+
+            // Here we could execute it in event loop async to prevent waiting io task on main
+            // stop reading new packets to prevent some packets came into pipeline too early
+            this.channel.config().setAutoRead(false);
+
+            // do our configuration task
+            final UnconfiguredPipelineHandler.OutboundConfigurationTask finalOutboundConfigurationTask = configMessage;
+            final Runnable writeTask = () -> this.channel.writeAndFlush(
+                    finalOutboundConfigurationTask.andThen(context -> this.sendLoginDisconnect = isLoginProtocol)
+            ).addListener(future -> {
+                try {
+                    if (future.isSuccess()) {
+                        if (callback != null) callback.run(); // retire callback if there have one
+                        return;
+                    }
+
+                    final Throwable ex = future.cause();
+
+                    // here we process our exceptions like that blocking one
+                    if (ex instanceof ClosedChannelException) {
+                        LOGGER.info("Connection closed during protocol change");
+                    } else {
+                        this.channel.pipeline().fireExceptionCaught(ex);
+                    }
+                }finally {
+                    // reset auto back and resume reading if needed
+                    if (resumeAutoReading) {
+                        this.channel.config().setAutoRead(true);
+                        this.channel.read(); // read once
+                    }
+                }
+            });
+
+            // we need to do this inside the event loop
+            if (!this.channel.eventLoop().inEventLoop()) {
+                this.channel.eventLoop().execute(writeTask);
+                return;
+            }
+
+            writeTask.run();
+        }
+    }
+    // Luminol end
+
     public <T extends PacketListener> void setupInboundProtocol(ProtocolInfo<T> protocolInfo, T packetInfo) {
         this.validateListener(protocolInfo, packetInfo);
         if (protocolInfo.flow() != this.getReceiving()) {
diff --git a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
index 2cc3dd1f267ae78d3aacb84509d4bf668235688f..00eec818bfd7d1a7442e6bd2fa531f4a4e872101 100644
--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -190,8 +190,9 @@ public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketLis
     public void handleConfigurationFinished(ServerboundFinishConfigurationPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.server.packetProcessor());
         this.finishCurrentTask(JoinWorldTask.TYPE);
-        this.connection.setupOutboundProtocol(GameProtocols.CLIENTBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess())));
+        // this.connection.setupOutboundProtocol(GameProtocols.CLIENTBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess()))); // Luminol - Async protocol switch - move down
 
+        Runnable afterSwitch = () -> { // Luminol - Async protocol switch
         try {
             PlayerList playerList = this.server.getPlayerList();
             if (playerList.getPlayer(this.gameProfile.id()) != null) {
@@ -237,6 +238,18 @@ public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketLis
             LOGGER.error("Couldn't place player in world", (Throwable)var4);
             this.disconnect(DISCONNECT_REASON_INVALID_DATA);
         }
+        // Luminol start - Async protocol switch
+        };
+        if (!me.earthme.luminol.config.modules.optimizations.AsyncProtocolChangeConfig.enabled) {
+            this.connection.setupOutboundProtocol(GameProtocols.CLIENTBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess())));
+            afterSwitch.run(); // directly run callback as we won't process any packet this time
+        } else {
+            this.connection.setupOutboundProtocolAsync(GameProtocols.CLIENTBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess())), () -> {
+                // push back
+                io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask(afterSwitch);
+            }, false); // we will start auto read once we set up inbound handler at placeNewPlayer in PlayerList
+        }
+        // Luminol end
     }
 
     @Override
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 4547f5f4588f1e0311b8008e06a711d9fdebc067..9dc3e00dd9f1110cbbe79e301a249d2420e0d94e 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2846,7 +2846,7 @@ public class ServerGamePacketListenerImpl
         // removed from the world
         // Folia end - rewrite login process - fix bad ordering of this field write - move after removed from world
         try { // Folia - rewrite login process - move connection ownership to global region
-            this.hackSwitchingConfig = true; // Folia - rewrite login process - avoid adding logout ticket here
+            this.hackSwitchingConfig = true; // Folia - rewrite login process - avoid adding logout ticket here and retiring scheduler
         this.removePlayerFromWorld();
         } finally { // Folia start - rewrite login process - move connection ownership to global region
             io.papermc.paper.threadedregions.RegionizedWorldData worldData = this.player.level().getCurrentWorldData();
@@ -2855,7 +2855,13 @@ public class ServerGamePacketListenerImpl
         } // Folia end - rewrite login process - move connection ownership to global region
         this.waitingForSwitchToConfig = true; // Folia - rewrite login process - fix bad ordering of this field write - moved down
         this.send(ClientboundStartConfigurationPacket.INSTANCE);
+        if (!me.earthme.luminol.config.modules.optimizations.AsyncProtocolChangeConfig.enabled) { // Luminol - Async protocol switch
         this.connection.setupOutboundProtocol(ConfigurationProtocols.CLIENTBOUND);
+        // Luminol start - Async protcol switch
+        } else {
+            this.connection.setupOutboundProtocolAsync(ConfigurationProtocols.CLIENTBOUND, null, true);
+        }
+        // Luminol end
     }
 
     @Override
@@ -3720,12 +3726,25 @@ public class ServerGamePacketListenerImpl
             throw new IllegalStateException("Client acknowledged config, but none was requested");
         } else {
             final ServerConfigurationPacketListenerImpl listener = new ServerConfigurationPacketListenerImpl(this.server, this.connection, this.createCookie(this.player.clientInformation())); // Paper
+            if (!me.earthme.luminol.config.modules.optimizations.AsyncProtocolChangeConfig.enabled){ // Luminol - Async protocol switch
             this.connection
                 .setupInboundProtocol(
                     ConfigurationProtocols.SERVERBOUND,
                     listener // Paper
                 );
-            new io.papermc.paper.event.connection.configuration.PlayerConnectionReconfigureEvent(listener.paperConnection).callEvent(); // Paper
+            new io.papermc.paper.event.connection.configuration.PlayerConnectionReconfigureEvent(listener.paperConnection).callEvent(); }// Paper // Luminol - Async protocol switch - add "{"
+            // Luminol start - Async protocol switch - move up
+            else
+                this.connection.setupInboundProtocolAsync(
+                        ConfigurationProtocols.SERVERBOUND,
+                        listener,
+                        () -> {
+                            new io.papermc.paper.event.connection.configuration.PlayerConnectionReconfigureEvent(listener.paperConnection).callEvent(); // Paper
+                        },
+                        true
+                );
+            // Luminol end
+            // new io.papermc.paper.event.connection.configuration.PlayerConnectionReconfigureEvent(listener.paperConnection).callEvent(); // Paper // Luminol - Async protocol switch - move up
         }
     }
 
diff --git a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index af111b5734cbdcfac6cbc7907f03acaa09294e36..43fddb41c666898493c01610b12060878574ab78 100644
--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -428,14 +428,35 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
     public void handleLoginAcknowledgement(ServerboundLoginAcknowledgedPacket packet) {
         net.minecraft.network.protocol.PacketUtils.ensureRunningOnSameThread(packet, this, this.server.packetProcessor()); // CraftBukkit
         Validate.validState(this.state == ServerLoginPacketListenerImpl.State.PROTOCOL_SWITCHING, "Unexpected login acknowledgement packet");
-        this.connection.setupOutboundProtocol(ConfigurationProtocols.CLIENTBOUND);
+        /*this.connection.setupOutboundProtocol(ConfigurationProtocols.CLIENTBOUND); // Luminol - Async protocol switch - Rewrite
         CommonListenerCookie commonListenerCookie = CommonListenerCookie.createInitial(Objects.requireNonNull(this.authenticatedProfile), this.transferred);
         ServerConfigurationPacketListenerImpl serverConfigurationPacketListenerImpl = new ServerConfigurationPacketListenerImpl(
             this.server, this.connection, commonListenerCookie
         );
         this.connection.setupInboundProtocol(ConfigurationProtocols.SERVERBOUND, serverConfigurationPacketListenerImpl);
         serverConfigurationPacketListenerImpl.startConfiguration();
+        this.state = ServerLoginPacketListenerImpl.State.ACCEPTED;*/ // Luminol - Async protocol switch - Rewrite
+
+        // Luminol start - Async protocol switch
         this.state = ServerLoginPacketListenerImpl.State.ACCEPTED;
+        final CommonListenerCookie commonListenerCookie = CommonListenerCookie.createInitial(Objects.requireNonNull(this.authenticatedProfile), this.transferred);
+        final ServerConfigurationPacketListenerImpl serverConfigurationPacketListenerImpl = new ServerConfigurationPacketListenerImpl(
+                this.server, this.connection, commonListenerCookie
+        );
+
+        Runnable afterSwitch = () -> io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask(serverConfigurationPacketListenerImpl::startConfiguration); // push back to main thread
+
+        if (!me.earthme.luminol.config.modules.optimizations.AsyncProtocolChangeConfig.enabled) {
+            this.connection.setupOutboundProtocol(ConfigurationProtocols.CLIENTBOUND);
+            this.connection.setupInboundProtocol(ConfigurationProtocols.SERVERBOUND, serverConfigurationPacketListenerImpl);
+            afterSwitch.run();
+            return;
+        }
+
+        this.connection.setupInboundProtocolAsync(ConfigurationProtocols.SERVERBOUND, serverConfigurationPacketListenerImpl, () -> {
+            this.connection.setupOutboundProtocolAsync(ConfigurationProtocols.CLIENTBOUND, afterSwitch, true); // start auto read when everything is ready
+        }, false); // we will resume auto reading once the outbound protocol is also setup
+        // Luminol end
     }
 
     @Override
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 8274afe0d07c50ece420b1d3cb20165ed0162acb..bbe7b262a3a9cbd10b286de5f7e8da1bd0f20b44 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -235,10 +235,12 @@ public abstract class PlayerList {
         // only after setting the connection listener to game type, add the connection to this regions list
         serverLevel.getCurrentWorldData().connections.add(connection);
         // Folia end - rewrite login process
+        if (!me.earthme.luminol.config.modules.optimizations.AsyncProtocolChangeConfig.enabled) { // Luminol - Async protocol switch // we will run async switch once these main thread logics became done
         connection.setupInboundProtocol(
             GameProtocols.SERVERBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess()), serverGamePacketListenerImpl),
             serverGamePacketListenerImpl
         );
+        } // Luminol - Async protocol switch
         serverGamePacketListenerImpl.suspendFlushing();
         GameRules gameRules = serverLevel.getGameRules();
         boolean flag = gameRules.get(GameRules.IMMEDIATE_RESPAWN);
@@ -401,6 +403,17 @@ public abstract class PlayerList {
             );
         }
         // Paper end - Send empty chunk
+        // Luminol start - Async protocol switch
+        if (me.earthme.luminol.config.modules.optimizations.AsyncProtocolChangeConfig.enabled) {
+            // auto read will be enabled once the async switch is done
+            connection.setupInboundProtocolAsync(
+                    GameProtocols.SERVERBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess()), serverGamePacketListenerImpl),
+                    serverGamePacketListenerImpl,
+                    null, // we don't need do anything more
+                    true // start auto read which we have disabled in configuration handler
+            );
+        }
+        // Luminol end
     }
 
     public void updateEntireScoreboard(ServerScoreboard scoreboard, ServerPlayer player) {
