From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mrhua269 <mrhua269@gmail.com>
Date: Tue, 11 Aug 2026 19:41:33 +0800
Subject: [PATCH] Leaf: Remove stream in entity visible effects filter


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 1cb8265c1313ad4893d45c344b03007fc4e2de4b..cacf4588d1a48aaf3986e330f95c0e4480105c37 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1034,12 +1034,15 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     private void updateSynchronizedMobEffectParticles() {
-        List<ParticleOptions> visibleEffectParticles = this.activeEffects
-            .values()
-            .stream()
-            .filter(MobEffectInstance::isVisible)
-            .map(MobEffectInstance::getParticleOptions)
-            .toList();
+        // Leaf start - Remove stream in entity visible effects filter
+        List<ParticleOptions> visibleEffectParticles = new java.util.ArrayList<>();
+
+        for (MobEffectInstance effect : this.activeEffects.values()) {
+            if (effect.isVisible()) {
+                visibleEffectParticles.add(effect.getParticleOptions());
+            }
+        }
+        // Leaf end - Remove stream in entity visible effects filter
         this.entityData.set(DATA_EFFECT_PARTICLES, visibleEffectParticles);
         this.entityData.set(DATA_EFFECT_AMBIENCE_ID, areAllEffectsAmbient(this.activeEffects.values()));
     }
