From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mrhua269 <mrhua269@gmail.com>
Date: Tue, 11 Aug 2026 19:48:57 +0800
Subject: [PATCH] Kaiiju: Vanilla end portal teleportation


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index a7e9d60da8a4f7babbd8bdcdaea9b19eae2320d0..8f3b139665c265efdb2e8fde86a6ec3604da743d 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4647,15 +4647,19 @@ public abstract class Entity
                         targetPos, 16, // load 16 blocks to be safe from block physics
                         ca.spottedleaf.concurrentutil.util.Priority.HIGH,
                         (chunks) -> {
-                            net.minecraft.world.level.levelgen.feature.EndPlatformFeature.createEndPlatform(destination, targetPos.below(), true, null);
-
+                            //net.minecraft.world.level.levelgen.feature.EndPlatformFeature.createEndPlatform(destination, targetPos.below(), true, null);  // Kaiiju - Vanilla end teleportation - moved down
+                            // Kaiiju start - Vanilla end teleportation
+                            Vec3 finalPos;
+                            if (this instanceof Player) finalPos = Vec3.atBottomCenterOf(targetPos.below());
+                            else finalPos = Vec3.atBottomCenterOf(targetPos);
+                            // Kaiiju end
                             // the portal obsidian is placed at targetPos.y - 2, so if we want to place the entity
                             // on the obsidian, we need to spawn at targetPos.y - 1
                             portalInfoCompletable.complete(
                                 new net.minecraft.world.level.portal.TeleportTransition(
-                                    destination, Vec3.atBottomCenterOf(targetPos.below()), Vec3.ZERO, Direction.WEST.toYRot(), 0.0f,
+                                    destination, finalPos, this.getDeltaMovement(), Direction.WEST.toYRot(), 0.0f, // Kaiiju - Vanilla end teleportation
                                     false, false,
-                                    Relative.union(Relative.DELTA, Set.of(Relative.X_ROT)),
+                                    /*Relative.union(Relative.DELTA, Set.of(Relative.X_ROT))*/Set.of(), // Kaiiju - Vanilla end teleportation
                                     TeleportTransition.PLAY_PORTAL_SOUND.then(TeleportTransition.PLACE_PORTAL_TICKET),
                                     org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_PORTAL,
                                     TeleportTransition.PassengerTeleportationMode.POSITION_RIDER
@@ -4682,11 +4686,15 @@ public abstract class Entity
                         ca.spottedleaf.concurrentutil.util.Priority.HIGH,
                         (chunks) -> {
                             BlockPos adjustedSpawn = destination.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, spawnPos);
-
+                            // Kaiiju start - Vanilla end teleportation
+                            Vec3 finalPos;
+                            if (this instanceof Player) finalPos = Vec3.atBottomCenterOf(adjustedSpawn.below());
+                            else finalPos = Vec3.atBottomCenterOf(adjustedSpawn);
+                            // Kaiiju end
                             // done
                             portalInfoCompletable.complete(
                                 new net.minecraft.world.level.portal.TeleportTransition(
-                                    destination, Vec3.atBottomCenterOf(adjustedSpawn), Vec3.ZERO, 0.0f, 0.0f,
+                                    destination, finalPos, this.getDeltaMovement(), 0.0f, 0.0f, // Kaiiju - Vanilla end teleportation
                                     false, false,
                                     Relative.union(Relative.DELTA, Relative.ROTATION),
                                     TeleportTransition.PLAY_PORTAL_SOUND.then(TeleportTransition.PLACE_PORTAL_TICKET),
@@ -4885,6 +4893,10 @@ public abstract class Entity
             return false;
         }
         // Luminol end
+        // Kaiiju start - sync end platform spawning & entity teleportation
+        final java.util.function.Consumer<Entity> tpComplete = type == PortalType.END && destination.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.END ?
+              e -> net.minecraft.world.level.levelgen.feature.EndPlatformFeature.createEndPlatform(destination, ServerLevel.END_SPAWN_POINT.below(), true, null) : teleportComplete;
+        // Kaiiju end
 
         Vec3 initialPosition = this.position();
         ChunkPos initialPositionChunk = new ChunkPos(
@@ -4949,9 +4961,14 @@ public abstract class Entity
                         info.postTeleportTransition().onTransition(teleported);
                     }
 
-                    if (teleportComplete != null) {
+                    // Kaiiju start - vanilla end teleportation
+                    /*if (teleportComplete != null) {
                         teleportComplete.accept(teleported);
+                    }*/
+                    if (tpComplete != null){
+                        tpComplete.accept(teleported);
                     }
+                    // Kaiiju end
                 }
             );
         });
