From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Tue, 11 Aug 2026 19:51:49 +0800
Subject: [PATCH] Pufferfish: Reduce chunk loading & lookups


diff --git a/net/minecraft/world/entity/monster/EnderMan.java b/net/minecraft/world/entity/monster/EnderMan.java
index 9b4ae8afda1f2202761cb49bc4f1181ce4eb55d5..82e1472fb39122dadef700b63cc784be71fd7a69 100644
--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -298,11 +298,18 @@ public class EnderMan extends Monster implements NeutralMob {
     private boolean teleport(final double x, final double y, final double z) {
         BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos(x, y, z);
 
-        while (pos.getY() > this.level().getMinY() && !this.level().getBlockState(pos).blocksMotion()) {
+        // Pufferfish start - single chunk lookup
+        net.minecraft.world.level.chunk.LevelChunk chunk = this.level().getChunkIfLoaded(pos);
+        if (chunk == null) {
+            return false;
+        }
+        // Pufferfish end
+        while (pos.getY() > this.level().getMinY() && !chunk.getBlockState(pos).blocksMotion()) { // Pufferfish
+
             pos.move(Direction.DOWN);
         }
 
-        BlockState blockState = this.level().getBlockState(pos);
+        BlockState blockState = chunk.getBlockState(pos); // Pufferfish
         boolean couldStandOn = blockState.blocksMotion();
         boolean isWet = blockState.getFluidState().is(FluidTags.WATER);
         if (couldStandOn && !isWet) {
