From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Tue, 11 Aug 2026 19:50:17 +0800
Subject: [PATCH] Petal: Reduce sensor work


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index d23cbf17d7261d6da6d99e3ae29aeb4759e8a763..82608f832cc2a77a017541636fe0b482842f156f 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -835,11 +835,12 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
             return;
         }
         // Paper end - Allow nerfed mobs to jump and float
+        int idBasedTickCount = this.tickCount + this.getId(); // Luminol - Petal - Move up
         ProfilerFiller profiler = Profiler.get();
         profiler.push("sensing");
-        this.sensing.tick();
+        if (idBasedTickCount % me.earthme.luminol.config.modules.optimizations.PetalReduceSensorWorkConfig.delayTicks == 0 || !me.earthme.luminol.config.modules.optimizations.PetalReduceSensorWorkConfig.enabled) this.sensing.tick(); // Luminol - Petal - Reduce sensor work
         profiler.pop();
-        int idBasedTickCount = this.tickCount + this.getId();
+        //int idBasedTickCount = this.tickCount + this.getId(); // Luminol - Petal - Move up
         if (idBasedTickCount % 2 != 0 && this.tickCount > 1) {
             profiler.push("targetSelector");
             this.targetSelector.tickRunningGoals(false);
