From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mrhua269 <mrhua269@gmail.com>
Date: Tue, 11 Aug 2026 19:53:30 +0800
Subject: [PATCH] Pufferfish: Throttle goal selector during inactive ticking


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 82608f832cc2a77a017541636fe0b482842f156f..b8fd825f42d3ce3081debc9bd7fd2ba6e8cb6cbb 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -216,11 +216,13 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     }
 
     // Paper start
+    int _pufferfish_inactiveTickDisableCounter = 0; // Pufferfish - throttle inactive goal selector ticking
     @Override
     public void inactiveTick() {
         super.inactiveTick();
         if (!this.aware) return; // Paper - Do not tick AI for inactive unaware mobs
-        if (this.goalSelector.inactiveTick()) {
+        boolean isThrottled = me.earthme.luminol.config.modules.optimizations.EntityGoalSelectorInactiveTickConfig.enabled && _pufferfish_inactiveTickDisableCounter++ % 20 != 0; // Pufferfish - throttle inactive goal selector ticking
+        if (this.goalSelector.inactiveTick() && !isThrottled) { // Pufferfish
             this.goalSelector.tick();
         }
         if (this.targetSelector.inactiveTick()) {
