From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Tue, 11 Aug 2026 19:57:01 +0800
Subject: [PATCH] Do not fire pre creature spawn event unless some plugin
 islistening it


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index 500dcbe924fca160bd673592e1aba384f6eb9ed2..8ca7d39933152cf335e38663700f3931baca5d49 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -391,16 +391,20 @@ public final class NaturalSpawner {
     ) {
         // Paper start - PreCreatureSpawnEvent
         EntityType<?> type = currentSpawnData.type();
-        com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
-                org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
-                org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(type), org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL
-        );
-        if (!event.callEvent()) {
-            if (event.shouldAbortSpawn()) {
-                return PreSpawnStatus.ABORT;
+        // Luminol start - Do not fire pre creature spawn event unless some plugin is listening it
+        if (com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent.getHandlerList().getRegisteredListeners().length != 0) {
+            com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
+                    org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
+                    org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(type), org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL
+            );
+            if (!event.callEvent()) {
+                if (event.shouldAbortSpawn()) {
+                    return PreSpawnStatus.ABORT;
+                }
+                return PreSpawnStatus.CANCELLED;
             }
-            return PreSpawnStatus.CANCELLED;
         }
+        // Luminol end - Do not fire pre creature spawn event unless some plugin is listening it
         final boolean success = type.getCategory() != MobCategory.MISC
         // Paper end - PreCreatureSpawnEvent
             && (type.canSpawnFarFromPlayer() || !(nearestPlayerDistanceSqr > type.getCategory().getDespawnDistance() * type.getCategory().getDespawnDistance()))
