From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Libalpm64 <Levamara@proton.me>
Date: Tue, 11 Aug 2026 21:06:18 +0800
Subject: [PATCH] Add item multitask config


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index a87e40c76156ae66a3fb9c7dee2e2461337b15f7..108cd2eddbcada496708c99755052b018c576558 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2189,7 +2189,11 @@ public class ServerGamePacketListenerImpl
                                 this.player.sendSpawnProtectionMessage(pos);
                             } else if (this.awaitingPositionFromClient == null && (level.mayInteract(this.player, pos) || passthroughSignInteraction)) {
                                 // Paper end - Allow using signs inside spawn protection
-                                this.player.stopUsingItem(); // CraftBukkit - SPIGOT-4706
+                                // Luminol start - Add item multitask config
+                                if (!me.earthme.luminol.config.modules.fixes.ItemMultitaskConfig.enabled) {
+                                    this.player.stopUsingItem(); // CraftBukkit - SPIGOT-4706
+                                }
+                                // Luminol end
                                 InteractionResult interactionResult = this.player.gameMode.useItemOn(this.player, level, itemStack, hand, blockHit);
                                 if (interactionResult.consumesAction()) {
                                     CriteriaTriggers.ANY_BLOCK_USE.trigger(this.player, blockHit.getBlockPos(), itemStack);
@@ -2388,9 +2392,13 @@ public class ServerGamePacketListenerImpl
                 return;
             }
             // CraftBukkit end
-            if (this.player.getInventory().getSelectedSlot() != packet.getSlot() && this.player.getUsedItemHand() == InteractionHand.MAIN_HAND) {
-                this.player.stopUsingItem();
+            // Luminol start - Add item multitask config
+            if (!me.earthme.luminol.config.modules.fixes.ItemMultitaskConfig.enabled) {
+                if (this.player.getInventory().getSelectedSlot() != packet.getSlot() && this.player.getUsedItemHand() == InteractionHand.MAIN_HAND) {
+                    this.player.stopUsingItem();
+                }
             }
+            // Luminol end
 
             this.player.getInventory().setSelectedSlot(packet.getSlot());
             this.player.resetLastActionTime();
