From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Tue, 11 Aug 2026 21:08:24 +0800
Subject: [PATCH] Leaves: Fakeplayer


diff --git a/io/papermc/paper/threadedregions/RegionizedWorldData.java b/io/papermc/paper/threadedregions/RegionizedWorldData.java
index f092ab76afd97760ca65f335e93cb33ab9385bdf..012e1a154407bd2e80324f110e4573be2102b114 100644
--- a/io/papermc/paper/threadedregions/RegionizedWorldData.java
+++ b/io/papermc/paper/threadedregions/RegionizedWorldData.java
@@ -408,7 +408,7 @@ public final class RegionizedWorldData {
     // Luminol end
     // connections
     private static final Connection[] EMPTY_CONNECTION_ARRAY = new Connection[0];
-    private final ReferenceList<Connection> connections = new ReferenceList<>(EMPTY_CONNECTION_ARRAY);
+    public final ReferenceList<Connection> connections = new ReferenceList<>(EMPTY_CONNECTION_ARRAY); // Lophine - BotList Public
 
     // misc. fields
     private boolean isHandlingTick;
diff --git a/net/minecraft/advancements/triggers/SimpleCriterionTrigger.java b/net/minecraft/advancements/triggers/SimpleCriterionTrigger.java
index d76a338ab4f326e6105adc451eee3d4601403c98..b827d51d23ceaebef2191a9f349bbd346e73f1af 100644
--- a/net/minecraft/advancements/triggers/SimpleCriterionTrigger.java
+++ b/net/minecraft/advancements/triggers/SimpleCriterionTrigger.java
@@ -17,6 +17,7 @@ import net.minecraft.world.level.storage.loot.ValidationContextSource;
 
 public abstract class SimpleCriterionTrigger<T extends SimpleCriterionTrigger.SimpleInstance> implements CriterionTrigger<T> {
     protected void trigger(final ServerPlayer player, final Predicate<T> matcher) {
+        if (player instanceof org.leavesmc.leaves.bot.ServerBot) return; // Leaves - bot skip
         PlayerAdvancements advancements = player.getAdvancements();
         Map<PlayerAdvancements.TriggerInstanceKey, T> listenersForType = advancements.getTriggerMapForType(this);
         if (listenersForType != null && !listenersForType.isEmpty()) {
diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index d2e9d46bbd9ee4ce45419b11499d7a7bc50f0219..848967dbe41694411603ef369cfdc3a0b00240d3 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -74,7 +74,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     public boolean preparing = true;
     // Spigot end
     private volatile @Nullable PacketListener disconnectListener;
-    private volatile @Nullable PacketListener packetListener;
+    protected volatile @Nullable PacketListener packetListener; // Leaves - private -> protected
     private @Nullable DisconnectionDetails disconnectionDetails;
     private final java.util.concurrent.atomic.AtomicBoolean disconnectionHandled = new java.util.concurrent.atomic.AtomicBoolean(false); // Folia - region threading - may be called concurrently during configuration stage
     private int receivedPackets;
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 9bbc8a6cd0178dd48951d8f44e4d2db97e8360e2..1aa738cc56590311351605b35180b6131859a4b3 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -296,6 +296,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     // Paper - per-level scheduledEvents
     private final ServerClockManager clockManager;
 
+    private org.leavesmc.leaves.bot.BotList botList; // Leaves - fakeplayer
+
     public static <S extends MinecraftServer> S spin(final Function<Thread, S> factory) {
         ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.init(); // Paper - rewrite data converter system
         AtomicReference<S> serverReference = new AtomicReference<>();
@@ -1191,6 +1193,16 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.halt(wait, false);
     }
     public void halt(final boolean wait, final boolean isRestarting) {
+        // Lophine start - Folia changes
+        halt(wait, isRestarting, false);
+    }
+
+    public void halt(final boolean wait, final boolean isRestarting, boolean force) {
+        if (!force && !this.getBotList().forceShutdown && !this.getBotList().removeAll()) { // Leaves - save or remove bot
+            this.getPlayerList().broadcastSystemMessage(Component.literal("Bot Still need to save, please wait! If you want to shuntdown without bot data saving, please use /stop force"), false);
+            return;
+        }
+        // Lophine end - Folia changes
         this.isRestarting = isRestarting;
         this.hasLoggedStop = true; // Paper - Debugging
         if (this.isDebugging()) io.papermc.paper.util.TraceUtil.dumpTraceForThread("Server stopped"); // Paper - Debugging
@@ -1655,7 +1667,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         int emptyTickThreshold = this.pauseWhenEmptySeconds() * 20;
         this.removeDisabledPluginsBlockingSleep(); // Paper - API to allow/disallow tick sleeping
         if (false && emptyTickThreshold > 0) { // Folia - region threading - this is complicated to implement, and even if done correctly is messy
-            if (this.playerList.getPlayerCount() == 0 && !this.tickRateManager.isSprinting() && this.pluginsBlockingSleep.isEmpty()) { // Paper - API to allow/disallow tick sleeping
+            if (this.playerList.getPlayerCount() == 0 && this.botList.bots.isEmpty() && !this.tickRateManager.isSprinting() && this.pluginsBlockingSleep.isEmpty()) { // Paper - API to allow/disallow tick sleeping // Leaves - fakeplayer
                 this.emptyTicks++;
             } else {
                 this.emptyTicks = 0;
@@ -1797,6 +1809,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             this.isSaving = true;
             if (playerSaveInterval > 0) {
                 this.playerList.saveAll(playerSaveInterval);
+                org.leavesmc.leaves.bot.BotList.INSTANCE.saveAllResumeBots(playerSaveInterval); // Leaves - resident fakeplayer
             }
             if (region == null && fullSave) { // Folia - region threading - don't auto save level data
                 this.saveGlobalData(false);
@@ -2047,6 +2060,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     protected void tickConnection() {
         this.getConnection().tick();
+        this.botList.networkTick(); // Leaves - fakeplayer
     }
 
     public void forceGameTimeSynchronization() {
@@ -3246,6 +3260,16 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         return this.debugSubscribers;
     }
 
+    // Leaves start - fakeplayer
+    protected void setBotList(org.leavesmc.leaves.bot.BotList botList) {
+        this.botList = botList;
+    }
+
+    public org.leavesmc.leaves.bot.BotList getBotList() {
+        return botList;
+    }
+    // Leaves end - fakeplayer
+
     public enum MultiplayerScope {
         OFF("off"),
         LAN("lan");
diff --git a/net/minecraft/server/PlayerAdvancements.java b/net/minecraft/server/PlayerAdvancements.java
index 66eb4142784322dbfe81bf78ff0ae57eb4a6bc50..c206c37d274f42d71bd9df55777a6950f0402192 100644
--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -166,6 +166,11 @@ public class PlayerAdvancements {
     }
 
     public boolean award(final AdvancementHolder holder, final String criterion) {
+        // Leaves start - bot can't get advancement
+        if (player instanceof org.leavesmc.leaves.bot.ServerBot) {
+            return false;
+        }
+        // Leaves end - bot can't get advancement
         boolean result = false;
         AdvancementProgress progress = this.getOrStartProgress(holder);
         boolean wasDone = progress.isDone();
diff --git a/net/minecraft/server/commands/BanIpCommands.java b/net/minecraft/server/commands/BanIpCommands.java
index dee5e12db878c13392083a46496de88d0b5924c3..922afd5e0fad3519ff2d7dc692574e5f65f119d6 100644
--- a/net/minecraft/server/commands/BanIpCommands.java
+++ b/net/minecraft/server/commands/BanIpCommands.java
@@ -40,6 +40,12 @@ public class BanIpCommands {
             return banIp(source, target, reason);
         } else {
             ServerPlayer player = source.getServer().getPlayerList().getPlayerByName(target);
+            // Leaves start - disable ban
+            if (player instanceof org.leavesmc.leaves.bot.ServerBot) {
+                source.sendFailure(Component.literal("Permission denied"));
+                return 0;
+            }
+            // Leaves end - disable ban
             if (player != null) {
                 return banIp(source, player.getIpAddress(), reason);
             } else {
diff --git a/net/minecraft/server/commands/BanPlayerCommands.java b/net/minecraft/server/commands/BanPlayerCommands.java
index c87bc86c57fde9d92e3927bdead5cdaf2e935744..950f8ed955bc09713348936ae7aa1a8b14006c2a 100644
--- a/net/minecraft/server/commands/BanPlayerCommands.java
+++ b/net/minecraft/server/commands/BanPlayerCommands.java
@@ -38,8 +38,15 @@ public class BanPlayerCommands {
     private static int banPlayers(final CommandSourceStack source, final Collection<NameAndId> players, final @Nullable Component reason) throws CommandSyntaxException {
         UserBanList list = source.getServer().getPlayerList().getBans();
         int count = 0;
-
+        boolean hasBot = false; // Leaves - disable kick
         for (NameAndId player : players) {
+            // Leaves start - disable ban
+            if (player.isBot()) {
+                source.sendFailure(Component.literal("Permission denied"));
+                hasBot = true;
+                continue;
+            }
+            // Leaves end - disable ban
             if (!list.isBanned(player)) {
                 UserBanListEntry entry = new UserBanListEntry(player, null, source.getTextName(), null, reason == null ? null : reason.getString());
                 list.add(entry);
@@ -53,7 +60,13 @@ public class BanPlayerCommands {
         }
 
         if (count == 0) {
-            throw ERROR_ALREADY_BANNED.create();
+            // Leaves start - disable kick
+            if (hasBot) {
+                return count;
+            } else {
+                throw ERROR_ALREADY_BANNED.create();
+            }
+            // Leaves end - disable kick
         } else {
             return count;
         }
diff --git a/net/minecraft/server/commands/KickCommand.java b/net/minecraft/server/commands/KickCommand.java
index bf361ef4cbf2a55b8bba5a75111257fb2d6d471f..82629ecca8b0e216db4ccb6f4c38de53ac1a6419 100644
--- a/net/minecraft/server/commands/KickCommand.java
+++ b/net/minecraft/server/commands/KickCommand.java
@@ -39,9 +39,17 @@ public class KickCommand {
             throw ERROR_SINGLEPLAYER.create();
         }
 
-        int count = 0;
+        boolean hasBot = false; // Leaves - disable kick
+            int count = 0;
 
         for (ServerPlayer player : players) {
+            // Leaves start - disable kick
+            if (player instanceof org.leavesmc.leaves.bot.ServerBot) {
+                source.sendFailure(Component.literal("Permission denied"));
+                hasBot = true;
+                continue;
+            }
+            // Leaves end - disable kick
             if (!source.getServer().isSingleplayerOwner(player.nameAndId())) {
                 player.connection.disconnect(reason, org.bukkit.event.player.PlayerKickEvent.Cause.KICKED); // Paper - kick event cause
                 source.sendSuccess(() -> Component.translatable("commands.kick.success", player.getDisplayName(), reason), true);
@@ -50,7 +58,13 @@ public class KickCommand {
         }
 
         if (count == 0) {
-            throw ERROR_KICKING_OWNER.create();
+            // Leaves start - disable kick
+            if (hasBot) {
+                return count;
+            } else {
+                throw ERROR_KICKING_OWNER.create();
+            }
+            // Leaves end - disable kick
         } else {
             return count;
         }
diff --git a/net/minecraft/server/commands/OpCommand.java b/net/minecraft/server/commands/OpCommand.java
index 636f56bfc0e9075b257e2eb14e858975bdedd801..20cbc0e65fea3fc2d0033259c2f35e158ea81926 100644
--- a/net/minecraft/server/commands/OpCommand.java
+++ b/net/minecraft/server/commands/OpCommand.java
@@ -39,6 +39,7 @@ public class OpCommand {
         int count = 0;
 
         for (NameAndId player : players) {
+            if (player.isBot()) continue; // Leaves - disable op
             if (!list.isOp(player)) {
                 list.op(player);
                 count++;
diff --git a/net/minecraft/server/commands/StopCommand.java b/net/minecraft/server/commands/StopCommand.java
index 3021b70b941ed09c5c58b98d30830a2279b8339c..d289f59475b617ae527d123239297f695468b80c 100644
--- a/net/minecraft/server/commands/StopCommand.java
+++ b/net/minecraft/server/commands/StopCommand.java
@@ -12,6 +12,10 @@ public class StopCommand {
             c.getSource().sendSuccess(() -> Component.translatable("commands.stop.stopping"), true);
             c.getSource().getServer().halt(false);
             return Command.SINGLE_SUCCESS;
-        }));
+        }).then(Commands.literal("force").requires(Commands.hasPermission(Commands.LEVEL_OWNERS)).executes(c -> {
+            c.getSource().sendSuccess(() -> Component.translatable("commands.stop.stopping"), true);
+            c.getSource().getServer().halt(false, false, true);
+            return 1;
+        })));
     }
 }
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 81934fb56744dbbb3fdd100e8ec83cb634034d04..f8d0af6473aa3e32408b85a868fb807d9da510e6 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -230,6 +230,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         }
 
         // Spigot start
+        this.setBotList(new org.leavesmc.leaves.bot.BotList(this)); // Leaves - fakeplayer
         this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage));
         org.spigotmc.SpigotConfig.init((java.io.File) this.options.valueOf("spigot-settings"));
         org.spigotmc.SpigotConfig.registerCommands();
@@ -253,6 +254,8 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         com.destroystokyo.paper.Metrics.PaperMetrics.startMetrics(); // Paper - start metrics
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // Paper - load version history now
 
+        this.getBotList().loadResumeBotInfo(); // Leaves - load resident bot info
+
         // this.worldData.setGameType(properties.gameMode.get()); // CraftBukkit - moved to world loading
         LOGGER.info("Default game type: {}", properties.gameMode.get());
         // Paper start - Unix domain socket support
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 33330c8bab53cb2c6719fe297ab86f69f5de6678..90b9b88b3cac82e1624df889e708448de6bf45e9 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -972,7 +972,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
         return this.level.moonrise$getEntityLookup().hasEntity(id); // Folia - region threading
     }
 
-    protected void addEntity(final Entity entity) {
+    public void addEntity(final Entity entity) {
         org.spigotmc.AsyncCatcher.catchOp("entity track"); // Spigot
         // Paper start - ignore and warn about illegal addEntity calls instead of crashing server
         if (!entity.valid || entity.level() != this.level || entity.moonrise$getTrackedEntity() != null) { // Folia - region threading
@@ -1282,6 +1282,13 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 ServerPlayer player = conn.getPlayer();
                 if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player)) {
                     this.removePlayer(player);
+                    // Leaves start - render bot
+                    if (entity instanceof org.leavesmc.leaves.bot.ServerBot bot) {
+                        if (bot.needSendFakeData(player)) {
+                            bot.sendFakeData(player.connection, false);
+                        }
+                    }
+                    // Leaves end - render bot
                 }
             }
         }
@@ -1413,6 +1420,13 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                     }
                 } else {
                     this.removePlayer(player);
+                    // Leaves start - render bot
+                    if (this.entity instanceof org.leavesmc.leaves.bot.ServerBot bot) {
+                        if (bot.needSendFakeData(player)) {
+                            bot.sendFakeData(player.connection, false);
+                        }
+                    }
+                    // Leaves end - render bot
                 }
             }
         }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 5e8137060d8c73d19bb1b6323dce0ae0de7046a2..ad50c337b7c5a10a72aae406af500c1f7fcf55d1 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -237,6 +237,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     private static final org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry();
     public final org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer persistentDataContainer = new org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer(DATA_TYPE_REGISTRY);
     //private final alternate.current.wire.WireHandler wireHandler = new alternate.current.wire.WireHandler(this); // Paper - optimize redstone (Alternate Current) // Folia - region threading - move to regionised world data
+    final List<ServerPlayer> realPlayers; // Leaves - skip
 
     @Override
     public @Nullable LevelChunk getChunkIfLoaded(int x, int z) {
@@ -704,6 +705,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         this.chunkDataController = new ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.ChunkDataController((ServerLevel)(Object)this, this.chunkTaskScheduler);
         // Paper end - rewrite chunk system
         this.getCraftServer().addWorld(this.getWorld()); // CraftBukkit
+        this.realPlayers = Lists.newArrayList(); // Leaves - skip
         this.updateTickData(); // Folia - region threading - make sure it is initialised before ticked
     }
 
@@ -2611,6 +2613,12 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         return this.players;
     }
 
+    // Leaves start - fakeplayer skip
+    public List<ServerPlayer> realPlayers() {
+        return this.realPlayers;
+    }
+    // Leaves end - fakeplayer skip
+
     @Override
     public void updatePOIOnBlockStateChange(final BlockPos pos, final BlockState oldState, final BlockState newState) {
         Optional<Holder<PoiType>> oldType = PoiTypes.forState(oldState);
@@ -3085,6 +3093,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             // ServerLevel.this.getChunkSource().addEntity(entity); // Paper - ignore and warn about illegal addEntity calls instead of crashing server; moved down below valid=true
             if (entity instanceof ServerPlayer player) {
                 ServerLevel.this.players.add(player);
+                // Leaves start - skip
+                if (!(player instanceof org.leavesmc.leaves.bot.ServerBot)) {
+                    ServerLevel.this.realPlayers.add(player);
+                }
+                // Leaves end - skip
                 if (player.isReceivingWaypoints()) {
                     ServerLevel.this.getWaypointManager().addPlayer(player);
                 }
@@ -3166,6 +3179,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             ServerLevel.this.getChunkSource().removeEntity(entity);
             if (entity instanceof ServerPlayer player) {
                 ServerLevel.this.players.remove(player);
+                // Leaves start - skip
+                if (!(player instanceof org.leavesmc.leaves.bot.ServerBot)) {
+                    ServerLevel.this.realPlayers.remove(player);
+                }
+                // Leaves end - skip
                 ServerLevel.this.getWaypointManager().removePlayer(player);
                 ServerLevel.this.updateSleepingPlayerList();
             }
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 44ae05b5c74d4d7502f4755dc3581918f839f3bc..df1dd758eedfd0998b3caab89e6fe92f6968bee5 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -232,7 +232,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     private static final boolean DEFAULT_SPAWN_EXTRA_PARTICLES_ON_FALL = false;
     public ServerGamePacketListenerImpl connection;
     private final MinecraftServer server;
-    public final ServerPlayerGameMode gameMode;
+    public ServerPlayerGameMode gameMode; // Leaves - not final
     private final PlayerAdvancements advancements;
     private final ServerStatsCounter stats;
     private float lastRecordedHealthAndAbsorption = Float.MIN_VALUE;
@@ -250,9 +250,9 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     private boolean canChatColor = true;
     private long lastActionTime = Util.getMillis();
     private @Nullable Entity camera;
-    private boolean isChangingDimension;
+    protected boolean isChangingDimension;
     public boolean seenCredits = false;
-    private final ServerRecipeBook recipeBook;
+    protected ServerRecipeBook recipeBook; // Leaves - not final and private -> protected
     private @Nullable Vec3 levitationStartPos;
     private int levitationStartTime;
     private boolean disconnected;
@@ -1471,7 +1471,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
     // Leaves end - exp fix
 
-    private void tellNeutralMobsThatIDied() {
+    protected void tellNeutralMobsThatIDied() { // Leaves private -> protected
         AABB aabb = new AABB(this.blockPosition()).inflate(32.0, 10.0, 32.0);
         this.level()
             .getEntitiesOfClass(Mob.class, aabb, EntitySelector.NO_SPECTATORS)
@@ -2092,6 +2092,11 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.lastSentHealth = -1.0F;
         this.lastSentFood = -1;
         this.teleportSpectators(transition, oldLevel);
+        // Leaves start - bot support
+        if (fun.bm.lophine.config.modules.function.FakeplayerConfig.enable) {
+            this.server.getBotList().bots.forEach(bot -> bot.sendFakeDataIfNeed(this, true)); // Leaves - render bot
+        }
+        // Leaves end - bot support
         // CraftBukkit start
         org.bukkit.event.player.PlayerChangedWorldEvent changeEvent = new org.bukkit.event.player.PlayerChangedWorldEvent(this.getBukkitEntity(), oldLevel.getWorld());
         this.level().getCraftServer().getPluginManager().callEvent(changeEvent);
diff --git a/net/minecraft/server/players/CachedUserNameToIdResolver.java b/net/minecraft/server/players/CachedUserNameToIdResolver.java
index ab6ae4c023d949cd88d4059c4575e4ab739bfc99..41f52ea5489c9a9ecbf73649e47ef4d7e8f5cf9d 100644
--- a/net/minecraft/server/players/CachedUserNameToIdResolver.java
+++ b/net/minecraft/server/players/CachedUserNameToIdResolver.java
@@ -122,6 +122,12 @@ public class CachedUserNameToIdResolver implements UserNameToIdResolver {
 
     @Override
     public Optional<NameAndId> get(final String name) {
+        // Leaves start - fix bot
+        org.leavesmc.leaves.bot.ServerBot bot = org.leavesmc.leaves.bot.BotList.INSTANCE.getBotByName(name);
+        if (bot != null) {
+            return Optional.of(bot.nameAndId());
+        }
+        // Leaves end - fix bot
         String userName = name.toLowerCase(Locale.ROOT);
         boolean stateLocked = true; try { this.stateLock.lock(); // Paper - Fix GameProfileCache concurrency
         CachedUserNameToIdResolver.GameProfileInfo profileInfo = this.profilesByName.get(userName);
diff --git a/net/minecraft/server/players/NameAndId.java b/net/minecraft/server/players/NameAndId.java
index 12190f2a2856b0d0cf95144d0b1622f1b9600bf6..be01f06d4ee02854345842c97382e8b6401ab948 100644
--- a/net/minecraft/server/players/NameAndId.java
+++ b/net/minecraft/server/players/NameAndId.java
@@ -8,18 +8,22 @@ import java.util.UUID;
 import net.minecraft.core.UUIDUtil;
 import org.jspecify.annotations.Nullable;
 
-public record NameAndId(UUID id, String name) {
+public record NameAndId(UUID id, String name, boolean isBot) { // Leaves - fakeplayer
     public static final Codec<NameAndId> CODEC = RecordCodecBuilder.create(
         i -> i.group(UUIDUtil.STRING_CODEC.fieldOf("id").forGetter(NameAndId::id), Codec.STRING.fieldOf("name").forGetter(NameAndId::name))
             .apply(i, NameAndId::new)
     );
 
     public NameAndId(final GameProfile profile) {
-        this(profile.id(), profile.name());
+        this(profile.id(), profile.name(), com.google.common.collect.Iterables.getFirst(profile.properties().get("is_bot"), "false").equals("true")); // Leaves - fakeplayer
     }
 
     public NameAndId(final com.mojang.authlib.yggdrasil.response.NameAndId profile) {
-        this(profile.id(), profile.name());
+        this(profile.id(), profile.name(), false); // Leaves - fakeplayer
+    }
+
+    public NameAndId(UUID uuid, String name) {
+        this(uuid, name, false); // Leaves - fakeplayer
     }
 
     public static @Nullable NameAndId fromJson(final JsonObject object) {
@@ -33,7 +37,7 @@ public record NameAndId(UUID id, String name) {
                 return null;
             }
 
-            return new NameAndId(uuid, object.get("name").getAsString());
+            return new NameAndId(uuid, object.get("name").getAsString(), false); // Leaves - fakeplayer
         } else {
             return null;
         }
@@ -46,7 +50,7 @@ public record NameAndId(UUID id, String name) {
 
     public static NameAndId createOffline(final String name) {
         UUID id = UUIDUtil.createOfflinePlayerUUID(name);
-        return new NameAndId(id, name);
+        return new NameAndId(id, name, false); // Leaves - fakeplayer
     }
 
     // Paper start - utility method for common conversion back to the game profile
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 75fda00984205153a767f9f26a8c6a33708c7f5c..1cb76221aef5b9730749c3aa2c9a1463fc81c3a9 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -317,6 +317,19 @@ public abstract class PlayerList {
             //return; // Folia - region threading - must still allow the player to connect, as we must add to chunk map before handling disconnect
         }
 
+        // Leaves start - bot support
+        if (fun.bm.mili.config.modules.function.FakeplayerConfig.enable) {
+            org.leavesmc.leaves.bot.ServerBot bot = this.server.getBotList().getBotByName(player.getScoreboardName());
+            if (bot != null) {
+                this.server.getBotList().removeBot(bot, org.leavesmc.leaves.event.bot.BotRemoveEvent.RemoveReason.INTERNAL, player.getBukkitEntity(), false, false);
+            }
+            this.server.getBotList().bots.forEach(bot1 -> {
+                bot1.sendPlayerInfo(player);
+                bot1.sendFakeDataIfNeed(player, true);
+            }); // Leaves - render bot
+        }
+        // Leaves end - bot support
+
         final net.kyori.adventure.text.Component jm = playerJoinEvent.joinMessage();
 
         if (jm != null && !jm.equals(net.kyori.adventure.text.Component.empty())) { // Paper - Adventure
@@ -786,6 +799,11 @@ public abstract class PlayerList {
             respawnReason
         ).callEvent();
         // Paper end
+        // Leaves start - bot support
+        if (fun.bm.lophine.config.modules.function.FakeplayerConfig.enable) {
+            this.server.getBotList().bots.forEach(bot -> bot.sendFakeDataIfNeed(player, true)); // Leaves - render bot
+        }
+        // Leaves end - bot support
 
         return player;
     }
@@ -888,11 +906,16 @@ public abstract class PlayerList {
 
     public String[] getPlayerNamesArray() {
         List<ServerPlayer> players = new java.util.ArrayList<>(this.players); // Folia - region threading
-        String[] names = new String[players.size()]; // Folia - region threading
+        String[] names = new String[players.size() + this.server.getBotList().bots.size()]; // Leaves - fakeplayer support
 
         for (int i = 0; i < players.size(); i++) { // Folia - region threading
             names[i] = players.get(i).getGameProfile().name(); // Folia - region threading
         }
+        // Leaves start - fakeplayer support
+        for (int i = this.players.size(); i < names.length; ++i) {
+            names[i] = this.server.getBotList().bots.get(i - this.players.size()).gameProfile.name();
+        }
+        // Leaves end - fakeplayer support
 
         return names;
     }
@@ -1009,8 +1032,15 @@ public abstract class PlayerList {
         }
     }
 
-    public @Nullable ServerPlayer getPlayerByName(final String name) {
-        return this.playersByName.get(name.toLowerCase(java.util.Locale.ROOT)); // Spigot
+    public @Nullable ServerPlayer getPlayerByName(String name) {
+        // Leaves start - fakeplayer support
+        name = name.toLowerCase(java.util.Locale.ROOT);
+        ServerPlayer player = this.playersByName.get(name);
+        if (player == null) {
+            player = this.server.getBotList().getBotByName(name);
+        }
+        return player; // Spigot
+        // Leaves end - fakeplayer support
     }
 
     public void broadcast(
@@ -1356,7 +1386,13 @@ public abstract class PlayerList {
     }
 
     public @Nullable ServerPlayer getPlayer(final UUID uuid) {
-        return this.playersByUUID.get(uuid);
+        // Leaves start - fakeplayer support
+        ServerPlayer player = this.playersByUUID.get(uuid);
+        if (player == null) {
+            player = this.server.getBotList().getBot(uuid);
+        }
+        return player;
+        // Leaves start - fakeplayer support
     }
 
     public @Nullable ServerPlayer getPlayer(final String playerName) {
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 2e238d1acee0d7189df5ff5949835978105d54bf..484d1460e6d6530ef32b2fbbd52263eada7c817f 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1260,7 +1260,7 @@ public abstract class Entity
 
             BlockPos effectPos = this.getOnPosLegacy();
             BlockState effectState = this.level().getBlockState(effectPos);
-            if (this.isLocalInstanceAuthoritative()) {
+            if (this.isLocalInstanceAuthoritative() || this instanceof org.leavesmc.leaves.bot.ServerBot) {  // Leaves - ServerBot needs check fall damage
                 this.checkFallDamage(movement.y, this.onGround(), effectState, effectPos);
             }
 
@@ -1662,7 +1662,7 @@ public abstract class Entity
         return colliders.isEmpty() ? maxDistance : -Shapes.collide(Direction.Axis.Y, aabb, colliders, -maxDistance);
     }
 
-    private Vec3 collide(final Vec3 movement) {
+    public Vec3 collide(final Vec3 movement) { // Leaves - private -> public
         // Paper start - optimise collisions
         final boolean xZero = movement.x == 0.0;
         final boolean yZero = movement.y == 0.0;
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index b1c1c49c98e4d1ffe7489321d76c1a4b701f055c..e78323d93a857c02081124f4d396ab22e6d9b4be 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1961,7 +1961,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return 0; // CraftBukkit
     }
 
-    protected void dropExperience(ServerLevel level, @Nullable Entity entity) {
+    public void dropExperience(ServerLevel level, @Nullable Entity entity) {
         // CraftBukkit start - Update getExpReward() above if the removed if() changes!
         if (!(this instanceof net.minecraft.world.entity.boss.enderdragon.EnderDragon)) { // CraftBukkit - SPIGOT-2420: Special case ender dragon will drop the xp over time
             ExperienceOrb.awardWithDirection(level, this.position(), Vec3.ZERO, this.expToDrop, this instanceof ServerPlayer ? org.bukkit.entity.ExperienceOrb.SpawnReason.PLAYER_DEATH : org.bukkit.entity.ExperienceOrb.SpawnReason.ENTITY_DEATH, entity, this); // Paper
@@ -3297,7 +3297,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     private void travelRidden(final Player controller, final Vec3 selfInput) {
         Vec3 riddenInput = this.getRiddenInput(controller, selfInput);
         this.tickRidden(controller, riddenInput);
-        if (this.canSimulateMovement()) {
+        if (this.canSimulateMovement() || this.getControllingPassenger() instanceof org.leavesmc.leaves.bot.ServerBot) { // Leaves - Fakeplayer
             this.setSpeed(this.getRiddenSpeed(controller));
             this.travel(riddenInput);
         } else {
@@ -4230,7 +4230,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         // Paper start - lag compensate eating
         // we add 1 to the expected time to avoid lag compensating when we should not
         final boolean shouldLagCompensate = this.useItem.has(DataComponents.FOOD) && this.eatStartTime != -1 && (System.nanoTime() - this.eatStartTime) > ((1L + this.totalEatTimeTicks) * 50L * (1000L * 1000L));
-        if ((--this.useItemRemaining == 0 || shouldLagCompensate) && !this.level().isClientSide() && !useItem.useOnRelease()) {
+        if ((--this.useItemRemaining == 0 || shouldLagCompensate) && !(this instanceof org.leavesmc.leaves.bot.ServerBot) && !this.level().isClientSide() && !useItem.useOnRelease()) {
             this.useItemRemaining = 0;
             // Paper end - lag compensate eating
             this.completeUsingItem();
@@ -4419,6 +4419,23 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         this.stopUsingItem();
     }
 
+    // Leaves start - Fakeplayer
+    public boolean releaseUsingItemWithResult() {
+        ItemStack itemInHand = this.getItemInHand(this.getUsedItemHand());
+        boolean result = false;
+        if (!this.useItem.isEmpty() && ItemStack.isSameItem(itemInHand, this.useItem)) {
+            this.useItem = itemInHand;
+            result = this.useItem.releaseUsingWithResult(this.level(), this, this.getUseItemRemainingTicks());
+            if (this.useItem.useOnRelease()) {
+                this.updatingUsingItem();
+            }
+        }
+
+        this.stopUsingItem();
+        return result;
+    }
+    // Leaves end - Fakeplayer
+
     public void stopUsingItem() {
         if (!this.level().isClientSide()) {
             boolean wasUsingItem = this.isUsingItem();
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index a26025ec6c3c6fe84bfc598176a6d4ff49de4a41..0eb7d100a97e72b52443927709e0fa9afc6b7d0d 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -156,7 +156,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     protected FoodData foodData = new FoodData();
     protected int jumpTriggerTime;
     public int takeXpDelay;
-    private int sleepCounter = 0;
+    protected int sleepCounter = 0;
     protected boolean wasUnderwater;
     private final Abilities abilities = new Abilities();
     public int experienceLevel = 0;
@@ -167,7 +167,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     private int lastLevelUpTime;
     public GameProfile gameProfile;
     private boolean reducedDebugInfo;
-    private ItemStack lastItemInMainHand = ItemStack.EMPTY;
+    protected ItemStack lastItemInMainHand = ItemStack.EMPTY; // Leaves - fakeplayer
     private final ItemCooldowns cooldowns = this.createItemCooldowns();
     private Optional<GlobalPos> lastDeathLocation = Optional.empty();
     public me.earthme.luminol.functions.bars.TickableStatusBarList statusBarList = new me.earthme.luminol.functions.bars.TickableStatusBarList(this); // Luminol status bars
@@ -305,6 +305,12 @@ public abstract class Player extends Avatar implements ContainerUser {
         this.updatePlayerPose();
     }
 
+    // Leaves start - fakeplayer
+    protected void livingEntityTick() {
+        super.tick();
+    }
+    // Leaves end - fakeplayer
+
     @Override
     protected float getMaxHeadRotationRelativeToBody() {
         return this.isBlocking() ? 15.0F : super.getMaxHeadRotationRelativeToBody();
@@ -530,7 +536,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     public void removeEntitiesOnShoulder() {
     }
 
-    private void touch(final Entity entity) {
+    public void touch(final Entity entity) { // Leaves - private -> public
         entity.playerTouch(this);
     }
 
@@ -1230,7 +1236,7 @@ public abstract class Player extends Avatar implements ContainerUser {
             // Paper end - Configurable sprint interruption on attack
         }
 
-        if (entity instanceof ServerPlayer serverPlayer && entity.hurtMarked) {
+        if ((entity instanceof ServerPlayer serverPlayer && !(entity instanceof org.leavesmc.leaves.bot.ServerBot)) && entity.hurtMarked) { // Leaves - bot knockback
             // CraftBukkit start - Add Velocity Event
             boolean cancelled = false;
             org.bukkit.entity.Player player = (org.bukkit.entity.Player) entity.getBukkitEntity();
diff --git a/net/minecraft/world/entity/projectile/FishingHook.java b/net/minecraft/world/entity/projectile/FishingHook.java
index d7217fd075c6d6fae3622c56e940e247008dcc27..0958e7f93494766e97e87ed4d39c8e4550a641f6 100644
--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -59,7 +59,7 @@ public class FishingHook extends Projectile {
     private static final EntityDataAccessor<Integer> DATA_HOOKED_ENTITY = SynchedEntityData.defineId(FishingHook.class, EntityDataSerializers.INT);
     private static final EntityDataAccessor<Boolean> DATA_BITING = SynchedEntityData.defineId(FishingHook.class, EntityDataSerializers.BOOLEAN);
     private int life;
-    private int nibble;
+    public int nibble; // Leaves - private -> public
     public int timeUntilLured;
     public int timeUntilHooked;
     public float fishAngle;
diff --git a/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java b/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
index 9b887b40236a212de513e3ceece1abfec5372ae2..2f8f8fa1bef9f333978b47c2ce729c291bec66fb 100644
--- a/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
+++ b/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
@@ -270,6 +270,11 @@ public abstract class AbstractBoat extends VehicleEntity implements Leashable {
             }
 
             this.move(MoverType.SELF, this.getDeltaMovement());
+        } else if (this.getControllingPassenger() instanceof org.leavesmc.leaves.bot.ServerBot) { // Leaves start - Fakeplayer
+            this.floatBoat();
+            this.controlBoat();
+            this.move(MoverType.SELF, this.getDeltaMovement());
+            // Leaves end - Fakeplayer
         } else {
             this.setDeltaMovement(Vec3.ZERO);
         }
@@ -380,6 +385,13 @@ public abstract class AbstractBoat extends VehicleEntity implements Leashable {
         }
     }
 
+    // Leaves start - Fakeplayer
+    @Override
+    public boolean canSimulateMovement() {
+        return super.canSimulateMovement() || this.getControllingPassenger() instanceof org.leavesmc.leaves.bot.ServerBot;
+    }
+    // Leaves end - Fakeplayer
+
     protected @Nullable SoundEvent getPaddleSound() {
         return switch (this.getStatus()) {
             case IN_WATER, UNDER_WATER, UNDER_FLOWING_WATER -> SoundEvents.BOAT_PADDLE_WATER;
diff --git a/net/minecraft/world/inventory/AbstractContainerMenu.java b/net/minecraft/world/inventory/AbstractContainerMenu.java
index c063694237c0c98ad370ac7fc170172f9ed3fb37..67812fa13df6a7537f658499677e19f38acddd8c 100644
--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -398,6 +398,7 @@ public abstract class AbstractContainerMenu {
 
     private void doClick(final int slotIndex, final int buttonNum, final ContainerInput containerInput, final Player player) {
         Inventory inventory = player.getInventory();
+        if (!doClickCheck(slotIndex, buttonNum, containerInput, player)) return; // Leaves - doClick check
         if (containerInput == ContainerInput.QUICK_CRAFT) {
             int expectedStatus = this.quickcraftStatus;
             this.quickcraftStatus = getQuickcraftHeader(buttonNum);
@@ -671,6 +672,22 @@ public abstract class AbstractContainerMenu {
         }
     }
 
+    // Leaves start - doClick check
+    private boolean doClickCheck(int slotIndex, int buttonNum, ContainerInput containerInput, Player player) {
+        if (slotIndex < 0) {
+            return true;
+        }
+
+        Slot slot = getSlot(slotIndex);
+        ItemStack itemStack = slot.getItem();
+        net.minecraft.world.item.component.CustomData customData = itemStack.get(net.minecraft.core.component.DataComponents.CUSTOM_DATA);
+        if (customData != null && customData.contains("Leaves.Gui.Placeholder")) {
+            return !customData.copyTag().getBoolean("Leaves.Gui.Placeholder").orElse(false);
+        }
+        return true;
+    }
+    // Leaves end - doClick check
+
     private boolean tryItemClickBehaviourOverride(
         final Player player, final ClickAction clickAction, final Slot slot, final ItemStack clicked, final ItemStack carried
     ) {
diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 5cbfc9e178b5823f92c0afedb1b0248d34c96292..dda6ad681cdd652ecaf186173492719c9459e3d3 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -454,7 +454,7 @@ public final class ItemStack implements DataComponentHolder, ItemInstance, Chang
                 placeEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockPlaceEvent(level, player, hand, capturedBlockStates.getFirst(), pos);
             }
 
-            if (placeEvent != null && (placeEvent.isCancelled() || !placeEvent.canBuild())) {
+            if (placeEvent != null && (placeEvent.isCancelled() || !placeEvent.canBuild()) && (!(player instanceof org.leavesmc.leaves.bot.ServerBot))) { // Leaves - Fakeplayer skip this check
                 result = InteractionResult.FAIL; // cancel placement
                 // PAIL: Remove this when MC-99075 fixed
                 player.containerMenu.forceHeldSlot(hand);
@@ -958,6 +958,20 @@ public final class ItemStack implements DataComponentHolder, ItemInstance, Chang
         }
     }
 
+    // Leaves start - Fakeplayer
+    public boolean releaseUsingWithResult(Level level, LivingEntity livingEntity, int timeLeft) {
+        ItemStack itemStack = this.copy();
+        if (this.getItem().releaseUsing(this, level, livingEntity, timeLeft)) {
+            ItemStack itemStack1 = this.applyAfterUseComponentSideEffects(livingEntity, itemStack);
+            if (itemStack1 != this) {
+                livingEntity.setItemInHand(livingEntity.getUsedItemHand(), itemStack1);
+            }
+            return true;
+        }
+        return false;
+    }
+    // Leaves end - Fakeplayer
+
     public boolean useOnRelease() {
         return this.getItem().useOnRelease(this);
     }
diff --git a/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java b/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
index 8d761640af4ce3afb4de5b46d3627614fff0ece9..15605dacaec3860e476a70dbdd3743e4197f67a6 100644
--- a/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+++ b/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
@@ -152,7 +152,7 @@ public class PistonMovingBlockEntity extends BlockEntity {
                                 break;
                             }
 
-                            if (!(entity instanceof ServerPlayer)) {
+                            if (!(entity instanceof ServerPlayer) || (entity instanceof org.leavesmc.leaves.bot.ServerBot)) { // Leaves - bot slime block
                                 Vec3 deltaMovement = entity.getDeltaMovement();
                                 double dx = deltaMovement.x;
                                 double dy = deltaMovement.y;
diff --git a/net/minecraft/world/level/levelgen/PhantomSpawner.java b/net/minecraft/world/level/levelgen/PhantomSpawner.java
index db0324595d7f5b6f92bd7724daf181b0903e1da4..ec3a437bd8a7d0e95e1e5fa0ff4462a70391ecf9 100644
--- a/net/minecraft/world/level/levelgen/PhantomSpawner.java
+++ b/net/minecraft/world/level/levelgen/PhantomSpawner.java
@@ -49,6 +49,10 @@ public class PhantomSpawner implements CustomSpawner {
                                         ServerStatsCounter stats = player.getStats();
                                         int value = Mth.clamp(stats.getValue(Stats.CUSTOM.get(Stats.TIME_SINCE_REST)), 1, Integer.MAX_VALUE);
                                         int dayLength = 24000;
+                                        if (player instanceof org.leavesmc.leaves.bot.ServerBot bot && bot.getConfigValue(org.leavesmc.leaves.bot.agent.Configs.SPAWN_PHANTOM)) {
+                                            dayLength = Math.max(bot.notSleepTicks, 1);
+                                        }
+                                        // Leaves end - fakeplayer spawn
                                         if (random.nextInt(value) >= level.paperConfig().entities.behavior.playerInsomniaStartTicks) { // Paper - Ability to control player's insomnia and phantoms
                                             BlockPos spawnPos = playerPos.above(20 + random.nextInt(15))
                                                 .east(-10 + random.nextInt(21))
