From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 11 Aug 2026 21:08:26 +0800
Subject: [PATCH] Leaves: BBOR Protocol


diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index a74eb2079863e4aaa8e982497c0fcd890d7de44b..1146728b922ae7381e56923502d9a868a727be17 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -755,6 +755,11 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
 
     public void setLoaded(final boolean loaded) {
         this.loaded = loaded;
+        // Leaves start - bbor
+        if (loaded) {
+            org.leavesmc.leaves.protocol.BBORProtocol.onChunkLoaded(this);
+        }
+        // Leaves end - bbor
     }
 
     public Level getLevel() {
