From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xucy10 <someonm20141923@outlook.com>
Date: Tue, 11 Aug 2026 21:15:04 +0800
Subject: [PATCH] Global-Entities-Counter


diff --git a/io/papermc/paper/threadedregions/RegionizedWorldData.java b/io/papermc/paper/threadedregions/RegionizedWorldData.java
index d3394076bb780947af59c07c2cbe512b32038ac9..5286be4a57de92674d7944c07389d8cd907d44fe 100644
--- a/io/papermc/paper/threadedregions/RegionizedWorldData.java
+++ b/io/papermc/paper/threadedregions/RegionizedWorldData.java
@@ -557,6 +557,9 @@ public final class RegionizedWorldData {
     }
     // Luminol end
 
+    public int uniqueId = -1; // Mili - Global Entities Counter
+    public boolean underGlobalEntitiesCounter = false; // Mili - Global Entities Counter
+
     public final TickRegions.TickRegionData regionData;
 
     public RegionizedWorldData(final ServerLevel world, final TickRegions.TickRegionData regionData) {
@@ -569,6 +572,13 @@ public final class RegionizedWorldData {
         this.turbo = new io.papermc.paper.redstone.RedstoneWireTurbo((RedStoneWireBlock)Blocks.REDSTONE_WIRE);
         this.regionData = regionData;
 
+        // Mili start - Global Entities Counter
+        if (fun.bm.mili.config.modules.experiment.GlobalEntitiesCounter.enabled) {
+            uniqueId = fun.bm.mili.utils.EntitiesCounterUtil.generateUniqueId();
+            underGlobalEntitiesCounter = true;
+        }
+        // Mili end - Global Entities Counter
+
         // tasks may be drained before the region ticks, so we must set up the tick data early just in case
         this.updateTickData();
     }
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index 414c21345879f589fb61130180d0ca606e37a454..e837b367c87b51ed0cc48ca525cba75038e9f574 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -511,6 +511,23 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         this.chunkMap.tick(haveTime);
         profiler.pop();
         this.clearCache();
+        // Mili start - Global Entities Counter
+        if (fun.bm.mili.config.modules.experiment.GlobalEntitiesCounter.enabled) {
+            io.papermc.paper.threadedregions.RegionizedWorldData regionizedWorldData = this.level.getCurrentWorldData();
+            if (regionizedWorldData != null) {
+                if (regionizedWorldData.underGlobalEntitiesCounter) {
+                    if (regionizedWorldData.getLocalPlayers().isEmpty()) {
+                        fun.bm.mili.utils.EntitiesCounterUtil.onWorldDataUnload(level, regionizedWorldData.uniqueId);
+                        regionizedWorldData.underGlobalEntitiesCounter = false;
+                    }
+                } else {
+                    if (!regionizedWorldData.getLocalPlayers().isEmpty()) {
+                        regionizedWorldData.underGlobalEntitiesCounter = true;
+                    }
+                }
+            }
+        }
+        // Mili end - Global Entities Counter
     }
 
     private void tickChunks() {
@@ -575,7 +592,13 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
             }
             spawnCookie = NaturalSpawner.createState(chunkCount, regionizedWorldData.getLoadedEntities(), this::getFullChunk, null, true); // Folia - region threading - note: function only cares about loaded entities, doesn't need all
         } else {
+            // Mili start - Global Entities Counter
+            if (fun.bm.mili.config.modules.experiment.GlobalEntitiesCounter.enabled) {
+                spawnCookie = fun.bm.mili.utils.EntitiesCounterUtil.runRemainingTasks(level, regionizedWorldData.getLoadedEntities(), this::getFullChunk, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false);
+            } else {
             spawnCookie = NaturalSpawner.createState(chunkCount, regionizedWorldData.getLoadedEntities(), this::getFullChunk, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false); // Folia - region threading - note: function only cares about loaded entities, doesn't need all
+            }
+            // Mili end - Global Entities Counter
         }
         } finally { foliaProfiler.stopTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.MOB_SPAWN_ENTITY_COUNT); } // Folia - profiler
         // Paper end - Optional per player mob spawns
@@ -583,7 +606,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         boolean doMobSpawning = this.level.getGameRules().get(GameRules.SPAWN_MOBS) && !this.level.getLocalPlayers().isEmpty(); // CraftBukkit // Folia - region threading
         int tickSpeed = this.level.getGameRules().get(GameRules.RANDOM_TICK_SPEED);
         List<MobCategory> spawningCategories;
-        if (doMobSpawning && (this.spawnEnemies || this.spawnFriendlies)) { // Paper
+        if (spawnCookie != null && doMobSpawning && (this.spawnEnemies || this.spawnFriendlies)) { // Paper // Mili - Global Entities Counter
             // Paper start - PlayerNaturallySpawnCreaturesEvent
             for (ServerPlayer player : this.level.getLocalPlayers()) { // Folia - region threading
                 int chunkRange = Math.min(level.spigotConfig.mobSpawnRange, player.getBukkitEntity().getViewDistance());
@@ -643,7 +666,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
             this.level.tickThunder(chunk);
         }
 
-        if (!spawningCategories.isEmpty()) {
+        if (spawnCookie != null && !spawningCategories.isEmpty()) { // Mili - Global Entities Counter
             if (this.level.getWorldBorder().isWithinBounds(chunkPos)) { // Paper - rewrite chunk system
                 NaturalSpawner.spawnForChunk(this.level, chunk, spawnCookie, spawningCategories);
             }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 8f742e14781e34b30d5d44b602e730031b24e594..508b58584d77aed0ddfc26e53dfa4198999575e4 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -817,6 +817,22 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
 
     public void tick(BooleanSupplier haveTime, final io.papermc.paper.threadedregions.TickRegions.TickRegionData region) { // Folia - regionised ticking
+        // Mili start - Global Entities Counter
+        if (fun.bm.mili.config.modules.experiment.GlobalEntitiesCounter.enabled) {
+            io.papermc.paper.threadedregions.RegionizedWorldData data = this.getCurrentWorldData();
+            if (data != null && data.underGlobalEntitiesCounter) {
+                fun.bm.mili.utils.EntitiesCounterUtil.tick(this, data.uniqueId, (ca.spottedleaf.moonrise.common.list.ReferenceList<Entity>) data.getLoadedEntities(), data.spawnChunkTracker);
+            }
+        }
+        // Mili end - Global Entities Counter
+        // Mili start - Cross Region Helper
+        if (fun.bm.mili.config.modules.experiment.CrossRegionHelperConfig.enabled) {
+            io.papermc.paper.threadedregions.RegionizedWorldData data = this.getCurrentWorldData();
+            if (data != null) {
+                fun.bm.mili.utils.CrossRegionHelper.onRegionTick(this, data);
+            }
+        }
+        // Mili end - Cross Region Helper
         final io.papermc.paper.threadedregions.RegionizedWorldData regionizedWorldData = this.getCurrentWorldData(); // Folia - regionised ticking
         final ca.spottedleaf.leafprofiler.RegionizedProfiler.Handle foliaProfiler = io.papermc.paper.threadedregions.TickRegionScheduler.getProfiler(); // Folia - profiler
         ProfilerFiller profiler = Profiler.get();
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 67e367f5dd4ecd706cd019a13c5e07e15a211573..1bf2cc3409ded38b65e6d2caafa481f6b442ac24 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -826,6 +826,44 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
                 }
             }
         } else {
+            // Mili start - fix some bug if player is cross scheduler >> to remove monster if the region don't have any valid player
+            if (fun.bm.mili.config.modules.experiment.GlobalEntitiesCounter.enabled) {
+                List<net.minecraft.server.level.ServerPlayer> playerList = this.level().getServer().getPlayerList().getPlayers();
+                if (!playerList.isEmpty()) {
+                    double d = -1.0;
+                    Vec3 vec3 = null;
+                    for (net.minecraft.server.level.ServerPlayer player1 : playerList) {
+                        if (player1.level() != this.level()) continue;
+                        double d1 = player1.distanceToSqr(this.position());
+                        if (d == -1.0 || d1 < d) {
+                            d = d1;
+                            vec3 = player1.position();
+                        }
+                    }
+                    if (vec3 != null) {
+                        // copy from PaperMC
+                        // Paper start - Configurable despawn distances
+                        final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory());
+                        final io.papermc.paper.configuration.type.DespawnRange.Shape shape = this.level().paperConfig().entities.spawning.despawnRangeShape;
+                        final double dy = Math.abs(vec3.y - this.getY());
+                        final double dySqr = Mth.square(dy);
+                        final double dxSqr = Mth.square(vec3.x - this.getX());
+                        final double dzSqr = Mth.square(vec3.z - this.getZ());
+                        final double distanceSquared = dxSqr + dzSqr + dySqr;
+                        // Despawn if hard/soft limit is exceeded
+                        if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
+                            this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                        }
+
+                        if (despawnRangePair.soft().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
+                            if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && this.removeWhenFarAway(distanceSquared)) {
+                                this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                            }
+                        }
+                    }
+                }
+            }
+            // Mili end
             this.noActionTime = 0;
         }
     }
diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index 8ca7d39933152cf335e38663700f3931baca5d49..e6488a0b6429f22bb6789e85867cde398e3456ae 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -121,7 +121,7 @@ public final class NaturalSpawner {
         return new NaturalSpawner.SpawnState(spawnableChunkCount, mobCounts, spawnPotential, localMobCapCalculator);
     }
 
-    private static Biome getRoughBiome(final BlockPos pos, final ChunkAccess chunk) {
+    public static Biome getRoughBiome(final BlockPos pos, final ChunkAccess chunk) { // Mili - private -> public
         return chunk.getNoiseBiome(QuartPos.fromBlock(pos.getX()), QuartPos.fromBlock(pos.getY()), QuartPos.fromBlock(pos.getZ())).value();
     }
 
@@ -629,7 +629,7 @@ public final class NaturalSpawner {
         private @Nullable EntityType<?> lastCheckedType;
         private double lastCharge;
 
-        private SpawnState(
+        public SpawnState( // Mili - private -> public
             final int spawnableChunkCount,
             final Object2IntOpenHashMap<MobCategory> mobCategoryCounts,
             final PotentialCalculator spawnPotential,
