From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xucy10 <someonm20141923@outlook.com>
Date: Tue, 21 Jul 2026 13:48:01 +0800
Subject: [PATCH] Global-Entities-Counter


diff --git a/io/papermc/paper/threadedregions/RegionizedWorldData.java b/io/papermc/paper/threadedregions/RegionizedWorldData.java
index b7ccdf5bf40603b929f768570ed232a6b27c0e9b..4a59842f80833456c6f2f793cbb1a662e3259347 100644
--- a/io/papermc/paper/threadedregions/RegionizedWorldData.java
+++ b/io/papermc/paper/threadedregions/RegionizedWorldData.java
@@ -566,6 +566,9 @@ public final class RegionizedWorldData {
     }
     // Luminol end
 
+    public int uniqueId = -1; // Mili - Global Entities Counter
+    public boolean underGlobalEntitiesCounter = false; // Mili - Global Entities Counter
+
     public final TickRegions.TickRegionData regionData;
 
     public RegionizedWorldData(final ServerLevel world, final TickRegions.TickRegionData regionData) {
@@ -578,6 +581,13 @@ public final class RegionizedWorldData {
         this.turbo = new io.papermc.paper.redstone.RedstoneWireTurbo((RedStoneWireBlock)Blocks.REDSTONE_WIRE);
         this.regionData = regionData;
 
+        // Mili start - Global Entities Counter
+        if (fun.bm.mili.config.modules.experiment.GlobalEntitiesCounter.enabled) {
+            uniqueId = fun.bm.mili.utils.EntitiesCounterUtil.generateUniqueId();
+            underGlobalEntitiesCounter = true;
+        }
+        // Mili end - Global Entities Counter
+
         // tasks may be drained before the region ticks, so we must set up the tick data early just in case
         this.updateTickData();
     }
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index f4c9c551fc39bae38e283ccbbc3cd116f0536a64..a58f785cbe4b5c18026f2e850f00448eb0307a0f 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -506,6 +506,23 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         this.chunkMap.tick(hasTimeLeft);
         profilerFiller.pop();
         this.clearCache();
+        // Mili start - Global Entities Counter
+        if (fun.bm.mili.config.modules.experiment.GlobalEntitiesCounter.enabled) {
+            io.papermc.paper.threadedregions.RegionizedWorldData regionizedWorldData = this.level.getCurrentWorldData();
+            if (regionizedWorldData != null) {
+                if (regionizedWorldData.underGlobalEntitiesCounter) {
+                    if (regionizedWorldData.getLocalPlayers().isEmpty()) {
+                        fun.bm.mili.utils.EntitiesCounterUtil.onWorldDataUnload(level, regionizedWorldData.uniqueId);
+                        regionizedWorldData.underGlobalEntitiesCounter = false;
+                    }
+                } else {
+                    if (!regionizedWorldData.getLocalPlayers().isEmpty()) {
+                        regionizedWorldData.underGlobalEntitiesCounter = true;
+                    }
+                }
+            }
+        }
+        // Mili end - Global Entities Counter
     }
 
     private void tickChunks() {
@@ -570,7 +587,13 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
             }
             spawnState = NaturalSpawner.createState(naturalSpawnChunkCount, regionizedWorldData.getLoadedEntities(), this::getFullChunk, null, true); // Folia - region threading - note: function only cares about loaded entities, doesn't need all
         } else {
+            // Mili start - Global Entities Counter
+            if (fun.bm.mili.config.modules.experiment.GlobalEntitiesCounter.enabled) {
+                spawnState = fun.bm.mili.utils.EntitiesCounterUtil.runRemainingTasks(level, regionizedWorldData.getLoadedEntities(), this::getFullChunk, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false);
+            } else {
             spawnState = NaturalSpawner.createState(naturalSpawnChunkCount, regionizedWorldData.getLoadedEntities(), this::getFullChunk, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false); // Folia - region threading - note: function only cares about loaded entities, doesn't need all
+            }
+            // Mili end - Global Entities Counter
         }
         } finally { foliaProfiler.stopTimer(ca.spottedleaf.leafprofiler.LProfilerRegistry.MOB_SPAWN_ENTITY_COUNT); } // Folia - profiler
         // Paper end - Optional per player mob spawns
@@ -578,7 +601,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         boolean flag = this.level.getGameRules().get(GameRules.SPAWN_MOBS) && !this.level.getLocalPlayers().isEmpty(); // CraftBukkit // Folia - region threading
         int i = this.level.getGameRules().get(GameRules.RANDOM_TICK_SPEED);
         List<MobCategory> filteredSpawningCategories;
-        if (flag && (this.spawnEnemies || this.spawnFriendlies)) { // Paper
+        if (spawnState != null && flag && (this.spawnEnemies || this.spawnFriendlies)) { // Paper // Mili - Global Entities Counter
             // Paper start - PlayerNaturallySpawnCreaturesEvent
             for (ServerPlayer entityPlayer : this.level.getLocalPlayers()) { // Folia - region threading
                 int chunkRange = Math.min(level.spigotConfig.mobSpawnRange, entityPlayer.getBukkitEntity().getViewDistance());
@@ -636,7 +659,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
             this.level.tickThunder(chunk);
         }
 
-        if (!spawnCategories.isEmpty()) {
+        if (spawnState != null && !spawnCategories.isEmpty()) { // Mili - Global Entities Counter
             if (this.level.getWorldBorder().isWithinBounds(pos)) { // Paper - rewrite chunk system
                 NaturalSpawner.spawnForChunk(this.level, chunk, spawnState, spawnCategories);
             }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index bf3b43c4db72a12879f938ac3850b2b7be7d219b..39419f1631fd6b49a661a006a8771a06810bd508 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -765,6 +765,22 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     }
 
     public void tick(BooleanSupplier hasTimeLeft, io.papermc.paper.threadedregions.TickRegions.TickRegionData region) { // Folia - regionised ticking
+        // Mili start - Global Entities Counter
+        if (fun.bm.mili.config.modules.experiment.GlobalEntitiesCounter.enabled) {
+            io.papermc.paper.threadedregions.RegionizedWorldData data = this.getCurrentWorldData();
+            if (data != null && data.underGlobalEntitiesCounter) {
+                fun.bm.mili.utils.EntitiesCounterUtil.tick(this, data.uniqueId, (ca.spottedleaf.moonrise.common.list.ReferenceList<Entity>) data.getLoadedEntities(), data.spawnChunkTracker);
+            }
+        }
+        // Mili end - Global Entities Counter
+        // Mili start - Cross Region Helper
+        if (fun.bm.mili.config.modules.experiment.CrossRegionHelperConfig.enabled) {
+            io.papermc.paper.threadedregions.RegionizedWorldData data = this.getCurrentWorldData();
+            if (data != null) {
+                fun.bm.mili.utils.CrossRegionHelper.onRegionTick(this, data);
+            }
+        }
+        // Mili end - Cross Region Helper
         final io.papermc.paper.threadedregions.RegionizedWorldData regionizedWorldData = this.getCurrentWorldData(); // Folia - regionised ticking
         final ca.spottedleaf.leafprofiler.RegionizedProfiler.Handle profiler = io.papermc.paper.threadedregions.TickRegionScheduler.getProfiler(); // Folia - profiler
         ProfilerFiller profilerFiller = Profiler.get();
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index c6985d2a9d1093943d754d77089bd4c3d5baf3ee..e6bb89efe4a54ba66d462774ceb236bae9005ee6 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -806,6 +806,44 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
                 }
             }
         } else {
+            // Mili start - fix some bug if player is cross scheduler >> to remove monster if the region don't have any valid player
+            if (fun.bm.mili.config.modules.experiment.GlobalEntitiesCounter.enabled) {
+                List<net.minecraft.server.level.ServerPlayer> playerList = this.level().getServer().getPlayerList().getPlayers();
+                if (!playerList.isEmpty()) {
+                    double d = -1.0;
+                    Vec3 vec3 = null;
+                    for (net.minecraft.server.level.ServerPlayer player1 : playerList) {
+                        if (player1.level() != this.level()) continue;
+                        double d1 = player1.distanceToSqr(this.position());
+                        if (d == -1.0 || d1 < d) {
+                            d = d1;
+                            vec3 = player1.position();
+                        }
+                    }
+                    if (vec3 != null) {
+                        // copy from PaperMC
+                        // Paper start - Configurable despawn distances
+                        final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory());
+                        final io.papermc.paper.configuration.type.DespawnRange.Shape shape = this.level().paperConfig().entities.spawning.despawnRangeShape;
+                        final double dy = Math.abs(vec3.y - this.getY());
+                        final double dySqr = Mth.square(dy);
+                        final double dxSqr = Mth.square(vec3.x - this.getX());
+                        final double dzSqr = Mth.square(vec3.z - this.getZ());
+                        final double distanceSquared = dxSqr + dzSqr + dySqr;
+                        // Despawn if hard/soft limit is exceeded
+                        if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
+                            this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                        }
+
+                        if (despawnRangePair.soft().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
+                            if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && this.removeWhenFarAway(distanceSquared)) {
+                                this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                            }
+                        }
+                    }
+                }
+            }
+            // Mili end
             this.noActionTime = 0;
         }
     }
diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index 91f0fbd75bc93355dc872390d91af83c76dc1dce..c5c14f787a74578d514ac909fcd673521ca1dc49 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -116,7 +116,7 @@ public final class NaturalSpawner {
         return new NaturalSpawner.SpawnState(spawnableChunkCount, map, potentialCalculator, calculator);
     }
 
-    static Biome getRoughBiome(BlockPos pos, ChunkAccess chunk) {
+    public static Biome getRoughBiome(BlockPos pos, ChunkAccess chunk) { // Mili - protected -> public
         return chunk.getNoiseBiome(QuartPos.fromBlock(pos.getX()), QuartPos.fromBlock(pos.getY()), QuartPos.fromBlock(pos.getZ())).value();
     }
 
@@ -585,7 +585,7 @@ public final class NaturalSpawner {
         private @Nullable EntityType<?> lastCheckedType;
         private double lastCharge;
 
-        SpawnState(
+        public SpawnState( // Mili - protected -> public
             int spawnableChunkCount,
             Object2IntOpenHashMap<MobCategory> mobCategoryCounts,
             PotentialCalculator spawnPotential,
