--- /dev/null
+++ b/net/minecraft/world/inventory/Slot.java.rej
@@ -1,0 +_,10 @@
+diff a/net/minecraft/world/inventory/Slot.java b/net/minecraft/world/inventory/Slot.java	(rejected hunks)
+@@ -75,7 +75,7 @@ public class Slot {
+     }
+ 
+     public int getMaxStackSize(ItemStack stack) {
+-        return Math.min(this.getMaxStackSize(), stack.getMaxStackSize());
++        return Math.min(this.getMaxStackSize(), org.leavesmc.leaves.util.ItemOverstackUtils.getItemStackMaxCount(stack)); // Mili - item over-stack util
+     }
+ 
+     public @Nullable Identifier getNoItemIcon() {
