--- /dev/null
+++ b/net/minecraft/world/level/ServerExplosion.java.rej
@@ -1,0 +_,24 @@
+diff a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java	(rejected hunks)
+@@ -545,7 +545,13 @@ public class ServerExplosion implements Explosion {
+                             // Paper end - knockback events
+                         }
+                         // CraftBukkit end
+-                        entity.push(vec32);
++                        // Leaves start - old projectile explosion behavior
++                        if (fun.bm.mili.config.modules.function.OldFeatureConfig.oldProjectileExplosionBehavior) {
++                            entity.setDeltaMovement(entity.getDeltaMovement().add(vec32));
++                        } else {
++                            entity.push(vec32);
++                        }
++                        // Leaves end - old projectile explosion behavior
+                         if (entity.getType().is(EntityTypeTags.REDIRECTABLE_PROJECTILE) && entity instanceof Projectile projectile) {
+                             projectile.setOwner(this.damageSource.getEntity());
+                         } else if (entity instanceof Player player && !player.isSpectator() && (!player.isCreative() || !player.getAbilities().flying) && !level.paperConfig().environment.disableExplosionKnockback) { // Paper - Option to disable explosion knockback
+@@ -721,6 +727,7 @@ public class ServerExplosion implements Explosion {
+     public boolean shouldAffectBlocklikeEntities() {
+         boolean flag = this.level.getGameRules().get(GameRules.MOB_GRIEFING);
+         boolean flag1 = this.source == null || this.source.getType() != EntityType.BREEZE_WIND_CHARGE && this.source.getType() != EntityType.WIND_CHARGE;
++        if (fun.bm.mili.config.modules.function.OldFeatureConfig.tntWetExplosionNoItemDamage) flag1 = flag1 && (this.source == null || !this.source.isInWater()); // Leaves - Old MC TNT wet explosion no item damage
+         return flag ? flag1 : this.blockInteraction.shouldAffectBlocklikeEntities() && flag1;
+     }
+ 
