--- /dev/null
+++ b/net/minecraft/world/SimpleContainer.java.rej
@@ -1,0 +_,19 @@
+diff a/net/minecraft/world/SimpleContainer.java b/net/minecraft/world/SimpleContainer.java	(rejected hunks)
+@@ -210,7 +210,7 @@ public class SimpleContainer implements Container, StackedContentsCompatible {
+     @Override
+     public void setItem(int index, ItemStack stack) {
+         this.items.set(index, stack);
+-        stack.limitSize(this.getMaxStackSize(stack));
++        stack.limitSize(this.getMaxStackLeaves(stack)); // Mili - item over-stack util
+         this.setChanged();
+     }
+ 
+@@ -285,7 +285,7 @@ public class SimpleContainer implements Container, StackedContentsCompatible {
+     }
+ 
+     private void moveItemsBetweenStacks(ItemStack stack, ItemStack other) {
+-        int maxStackSize = this.getMaxStackSize(other);
++        int maxStackSize = this.getMaxStackLeaves(other); // Mili - item over-stack util
+         int min = Math.min(stack.getCount(), maxStackSize - other.getCount());
+         if (min > 0) {
+             other.grow(min);
