From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xucy10 <xucy10@example.com>
Date: Thu, 10 Jul 2026 00:00:00 +0800
Subject: [PATCH] Rebrand to Mili


diff --git a/src/main/java/me/earthme/luminol/config/ConfigManager.java b/src/main/java/me/earthme/luminol/config/ConfigManager.java
index f5fb4e210c07f32e52eb8c1ba166d1571a6778c7..afa7df571dc2adaf76020ef20ea794ebf2194692 100644
--- a/src/main/java/me/earthme/luminol/config/ConfigManager.java
+++ b/src/main/java/me/earthme/luminol/config/ConfigManager.java
@@ -25,6 +25,19 @@ public class ConfigManager {
         configfiles.put("luminol", ConfigsInstance.of("luminol", "me.earthme.luminol.config.modules"));
         configfiles.put("lophine", ConfigsInstance.of("lophine", "fun.bm.lophine.config.modules")); // add lophine global config
+        configfiles.put("mili", ConfigsInstance.of(
+                new java.io.File("mili_config"),
+                "mili",
+                "mili_config.toml",
+                "miliconfig",
+                "fun.bm.mili.config.modules"
+        )); // add Mili global config
+        configfiles.put("mili_carpet", ConfigsInstance.of(
+                new java.io.File("mili_config"),
+                "mili_carpet",
+                "mili_carpet_config.toml",
+                "milicarpetconfig",
+                "fun.bm.mili.carpet.config.modules"
+        )); // add mili carpet config
         preLoad();
     }
 
diff --git a/src/main/java/me/earthme/luminol/config/modules/misc/ServerModNameConfig.java b/src/main/java/me/earthme/luminol/config/modules/misc/ServerModNameConfig.java
index a4188ef831825764c0756b826651e2e7a9f99c5b..1b4f1faba9740c39ae439f15cc45fece1b22a78e 100644
--- a/src/main/java/me/earthme/luminol/config/modules/misc/ServerModNameConfig.java
+++ b/src/main/java/me/earthme/luminol/config/modules/misc/ServerModNameConfig.java
@@ -8,7 +8,7 @@ import me.earthme.luminol.enums.EnumConfigCategory;
 @ConfigClassInfo(category = EnumConfigCategory.MISC, name = "server_mod_name")
 public class ServerModNameConfig implements IConfigModule {
     @ConfigInfo(name = "name", comments = "Decides the server mod name shown in your F3 debug screen.")
-    public static String serverModName = "Lophine";
+    public static String serverModName = "Mili";
 
     @ConfigInfo(name = "vanilla_spoof", comments = "Ignore any plugin's modification and server mod name set in this config block, only force sending brand name of vanilla")
     public static boolean fakeVanilla = false;