--- a/folia-server/build.gradle.kts
+++ b/folia-server/build.gradle.kts
@@ -7,7 +_,7 @@
     `java-library`
     `maven-publish`
     idea
-    id("io.papermc.paperweight.core")
+    id("moe.luminolmc.hyacinthusweight.core")
     id("io.papermc.fill.gradle") version "1.0.12"
 }
 
@@ -15,7 +_,7 @@
 
 dependencies {
     mache("io.papermc:mache:26.2+build.1")
-    paperclip("io.papermc:paperclip:3.0.4")
+    hyacinthusclip("moe.luminolmc:hyacinthusclip:${providers.gradleProperty("clipVersion").get()}")
 }
 
 paperweight {
@@ -24,6 +_,7 @@
     gitFilePatches = false
 
     val fork = forks.register("folia") {
+        rootDirectory = upstreamsDirectory().map { it.dir("folia") }
         upstream.patchDir("paperServer") {
             upstreamPath = "paper-server"
             excludes = setOf("src/minecraft", "patches", "build.gradle.kts")
@@ -32,7 +_,23 @@
         }
     }
 
-    activeFork = fork
+    val luminol = forks.register("luminol") {
+        forks = fork
+        upstream.patchRepo("paperServer") {
+            upstreamRepo = fork.patchedRepo("paperServer")
+            patchesDir = rootDirectory.dir("luminol-server/paper-patches")
+            outputDir = rootDirectory.dir("paper-server")
+        }
+
+        upstream.patchDir("foliaServer") {
+            upstreamPath = "folia-server"
+            excludes = setOf("src/minecraft", "paper-patches", "minecraft-patches", "build.gradle.kts", "build.gradle.kts.patch")
+            patchesDir = rootDirectory.dir("luminol-server/folia-patches")
+            outputDir = rootDirectory.dir("folia-server")
+        }
+    }
+
+    activeFork = luminol
 
 
     updatingMinecraft {
@@ -103,10 +_,14 @@
     main {
         java { srcDir("../paper-server/src/main/java"); srcDir("../paper-server/src/generated/java") }
         resources { srcDir("../paper-server/src/main/resources") }
+        java { srcDir("../folia-server/src/main/java") }
+        resources { srcDir("../folia-server/src/main/resources") }
     }
     test {
         java { srcDir("../paper-server/src/test/java") }
         resources { srcDir("../paper-server/src/test/resources") }
+        java { srcDir("../folia-server/src/test/java") }
+        resources { srcDir("../folia-server/src/test/resources") }
     }
 }
 val log4jPlugins = sourceSets.create("log4jPlugins") {
@@ -134,7 +_,22 @@
 }
 
 dependencies {
-    implementation(project(":folia-api"))
+    implementation(project(":mili-api")) // Mili
+    // Luminol start - Dependenices insert
+    implementation("net.objecthunter:exp4j:0.4.8")
+    implementation("io.netty:netty-all:4.2.9.Final") // used for io_uring
+    implementation("com.electronwill.night-config:toml:3.8.3") // Night config
+    implementation("com.github.luben:zstd-jni:1.5.4-1")
+    implementation("net.openhft:zero-allocation-hashing:0.16")
+    implementation("io.github.classgraph:classgraph:4.8.158")
+    implementation("net.openhft:affinity:3.23.3")
+    // Leaves start - leaves plugin
+    implementation("org.spongepowered:configurate-gson:4.2.0-SNAPSHOT") {
+        exclude(group = "com.google.code.gson", module = "gson")
+        exclude(group = "com.google.guava", module = "guava")
+    }
+    // Leaves end - leaves plugin
+    // Luminol end
     implementation("ca.spottedleaf:leafpile:1.0.0")
     implementation("org.jline:jline-terminal-ffm:3.27.1") // use ffm on java 22+
     implementation("org.jline:jline-terminal-jni:3.27.1") // fall back to jni on java 21
@@ -152,7 +_,7 @@
     runtimeOnly(log4jPlugins.output)
     alsoShade(log4jPlugins.output)
 
-    implementation("com.velocitypowered:velocity-native:3.4.0-SNAPSHOT") {
+    implementation("com.velocitypowered:velocity-native:3.4.0-SNAPSHOT") { // VelocityNT ReastLib
         isTransitive = false
     }
     implementation("io.netty:netty-codec-haproxy:4.2.15.Final") // Add support for proxy protocol
@@ -184,26 +_,43 @@
     implementation("me.lucko:spark-paper:1.10.152")
 }
 
+// Pufferfish Start
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+// Pufferfish End
+
+// Luminol start - Hide unnecessary warnings
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("-Xlint:-module")
+    compilerArgs.add("-Xlint:-removal")
+    compilerArgs.add("-Xlint:-dep-ann")
+}
+// Luminol end
+
+
 tasks.jar {
     manifest {
         val git = Git(rootProject.layout.projectDirectory.path)
         val mcVersion = rootProject.providers.gradleProperty("mcVersion").get()
         val build = System.getenv("BUILD_NUMBER") ?: null
-        val buildTime = providers.environmentVariable("BUILD_STARTED_AT").map(Instant::parse).orElse(Instant.EPOCH).get()
+        val buildTime = Instant.now() // Always use current as build time
         val gitHash = git.exec(providers, "rev-parse", "--short=7", "HEAD").get().trim()
         val implementationVersion = "$mcVersion-${build ?: "DEV"}-$gitHash"
         val date = git.exec(providers, "show", "-s", "--format=%ci", gitHash).get().trim()
         val gitBranch = git.exec(providers, "rev-parse", "--abbrev-ref", "HEAD").get().trim()
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
-            "Implementation-Title" to "Folia",
+            "Implementation-Title" to "Mili",
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date,
-            "Specification-Title" to "Folia",
+            "Specification-Title" to "Mili",
             "Specification-Version" to project.version,
-            "Specification-Vendor" to "Paper Team",
-            "Brand-Id" to "papermc:folia",
-            "Brand-Name" to "Folia",
+            "Specification-Vendor" to "LuminolMC org",
+            "Brand-Id" to "luminolmc:mili",
+            "Brand-Name" to "Mili",
             "Build-Number" to (build ?: ""),
             "Build-Time" to buildTime.toString(),
             "Git-Branch" to gitBranch,
@@ -352,7 +_,7 @@
 }
 
 fill {
-    project("folia")
+    project("mili")
     versionFamily(paperweight.minecraftVersion.map { it.split(".", "-").takeWhile { part -> part.toIntOrNull() != null }.take(2).joinToString(".") })
     version(paperweight.minecraftVersion)
 
