--- /dev/null
+++ b/net/minecraft/world/entity/animal/equine/AbstractHorse.java.rej
@@ -1,0 +_,15 @@
+diff a/net/minecraft/world/entity/animal/equine/AbstractHorse.java b/net/minecraft/world/entity/animal/equine/AbstractHorse.java	(rejected hunks)
+@@ -301,6 +301,13 @@ public abstract class AbstractHorse extends Animal implements HasCustomInventory
+     public void createInventory() {
+         SimpleContainer simpleContainer = this.inventory;
+         this.inventory = new SimpleContainer(this.getInventorySize(), (org.bukkit.entity.AbstractHorse) this.getBukkitEntity()); // CraftBukkit
++        // Leaves start - pca
++        this.inventory.addListener(inv -> {
++            if (fun.bm.mili.config.modules.function.protocol.PcaSyncProtocolConfig.enable) {
++                org.leavesmc.leaves.protocol.PcaSyncProtocol.syncEntityToClient(this);
++            }
++        });
++        // Leaves end - pca
+         if (simpleContainer != null) {
+             int min = Math.min(simpleContainer.getContainerSize(), this.inventory.getContainerSize());
+ 
