--- /dev/null
+++ b/net/minecraft/world/level/block/CactusBlock.java.rej
@@ -1,0 +_,12 @@
+diff a/net/minecraft/world/level/block/CactusBlock.java b/net/minecraft/world/level/block/CactusBlock.java	(rejected hunks)
+@@ -47,6 +47,10 @@ public class CactusBlock extends Block {
+     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+         if (!state.canSurvive(level, pos)) {
+             level.destroyBlock(pos, true);
++            // Leaves start - zero tick plants
++        } else if (fun.bm.mili.config.modules.function.OldFeatureConfig.zeroTickPlants) {
++            this.randomTick(state, level, pos, random);
++            // Leaves end - zero tick plants
+         }
+     }
+ 
