--- /dev/null
+++ b/net/minecraft/world/level/block/CopperBulbBlock.java.rej
@@ -1,0 +_,46 @@
+diff a/net/minecraft/world/level/block/CopperBulbBlock.java b/net/minecraft/world/level/block/CopperBulbBlock.java	(rejected hunks)
+@@ -33,14 +33,26 @@ public class CopperBulbBlock extends Block {
+     @Override
+     protected void onPlace(BlockState state, Level level, BlockPos pos, BlockState oldState, boolean movedByPiston) {
+         if (oldState.getBlock() != state.getBlock() && level instanceof ServerLevel serverLevel) {
+-            this.checkAndFlip(state, serverLevel, pos);
++            // Mili start - copper bulb 1 gt delay
++            if (!fun.bm.mili.config.modules.function.OldFeatureConfig.copperBulb1gt) {
++                this.checkAndFlip(state, serverLevel, pos);
++            } else {
++                level.scheduleTick(pos, this, 1);
++            }
++            // Mili end - copper bulb 1 gt delay
+         }
+     }
+ 
+     @Override
+     protected void neighborChanged(BlockState state, Level level, BlockPos pos, Block neighborBlock, @Nullable Orientation orientation, boolean movedByPiston) {
+         if (level instanceof ServerLevel serverLevel) {
+-            this.checkAndFlip(state, serverLevel, pos);
++            // Mili start - copper bulb 1 gt delay
++            if (!fun.bm.mili.config.modules.function.OldFeatureConfig.copperBulb1gt) {
++                this.checkAndFlip(state, serverLevel, pos);
++            } else {
++                level.scheduleTick(pos, this, 1);
++            }
++            // Mili end - copper bulb 1 gt delay
+         }
+     }
+ 
+@@ -57,6 +69,15 @@ public class CopperBulbBlock extends Block {
+         }
+     }
+ 
++    // Mili start - copper bulb 1 gt delay
++    @Override
++    protected void tick(BlockState state, ServerLevel world, BlockPos pos, net.minecraft.util.RandomSource random) {
++        if (fun.bm.mili.config.modules.function.OldFeatureConfig.copperBulb1gt) {
++            checkAndFlip(state, world, pos);
++        }
++    }
++    // Mili end - copper bulb 1 gt delay
++
+     @Override
+     protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
+         builder.add(LIT, POWERED);
