--- /dev/null
+++ b/net/minecraft/world/level/block/EndGatewayBlock.java.rej
@@ -1,0 +_,19 @@
+diff a/net/minecraft/world/level/block/EndGatewayBlock.java b/net/minecraft/world/level/block/EndGatewayBlock.java	(rejected hunks)
+@@ -107,6 +107,17 @@ public class EndGatewayBlock extends BaseEntityBlock implements Portal {
+     public @Nullable TeleportTransition getPortalDestination(ServerLevel level, Entity entity, BlockPos pos) {
+         if (level.getBlockEntity(pos) instanceof TheEndGatewayBlockEntity theEndGatewayBlockEntity) {
+             Vec3 portalPosition = theEndGatewayBlockEntity.getPortalPosition(level, pos);
++
++            // Mili start - force void trade
++            if (fun.bm.mili.config.modules.function.VillagerTradeConfig.forceVoidTrade && portalPosition != null && entity instanceof net.minecraft.server.level.ServerPlayer player) {
++                if (player.containerMenu instanceof net.minecraft.world.inventory.MerchantMenu merchantMenu) {
++                    if (merchantMenu.trader instanceof net.minecraft.world.entity.npc.villager.AbstractVillager villager) {
++                        villager.setVoidTrade();
++                    }
++                }
++            }
++            // Mili end - force void trade
++
+             if (portalPosition == null) {
+                 return null;
+             } else {
