--- /dev/null
+++ b/net/minecraft/world/level/block/SculkSensorBlock.java.rej
@@ -1,0 +_,26 @@
+diff a/net/minecraft/world/level/block/SculkSensorBlock.java b/net/minecraft/world/level/block/SculkSensorBlock.java	(rejected hunks)
+@@ -96,7 +96,7 @@ public class SculkSensorBlock extends BaseEntityBlock implements SimpleWaterlogg
+     @Override
+     public void stepOn(Level level, BlockPos pos, BlockState state, Entity entity) {
+         if (!level.isClientSide()
+-            && canActivate(state)
++            && canActivate(state, pos) // Mili - Sound update suppression
+             && entity.getType() != EntityType.WARDEN
+             && level.getBlockEntity(pos) instanceof SculkSensorBlockEntity sculkSensorBlockEntity
+             && level instanceof ServerLevel serverLevel
+@@ -211,6 +211,15 @@ public class SculkSensorBlock extends BaseEntityBlock implements SimpleWaterlogg
+         return state.getValue(PHASE);
+     }
+ 
++    // Mili start - Sound update suppression
++    public static boolean canActivate(BlockState state, BlockPos pos) {
++        if (fun.bm.mili.config.modules.function.OldFeatureConfig.soundUpdateSuppression && !(state.getBlock() instanceof SculkSensorBlock)) {
++            throw new org.leavesmc.leaves.util.UpdateSuppressionException(pos, null, state.getBlock(), null, new IllegalArgumentException());
++        }
++        return canActivate(state);
++    }
++    // Mili end - Sound update suppression
++
+     public static boolean canActivate(BlockState state) {
+         return state.getBlock() instanceof SculkSensorBlock &&  getPhase(state) == SculkSensorPhase.INACTIVE; // Paper - Check for a valid type
+     }
