--- /dev/null
+++ b/net/minecraft/world/level/levelgen/feature/CoralMushroomFeature.java.rej
@@ -1,0 +_,10 @@
+diff a/net/minecraft/world/level/levelgen/feature/CoralMushroomFeature.java b/net/minecraft/world/level/levelgen/feature/CoralMushroomFeature.java	(rejected hunks)
+@@ -14,7 +14,7 @@ public class CoralMushroomFeature extends CoralFeature {
+     }
+ 
+     @Override
+-    protected boolean placeFeature(LevelAccessor level, RandomSource random, BlockPos pos, BlockState state) {
++    public boolean placeFeature(LevelAccessor level, RandomSource random, BlockPos pos, BlockState state) { // Mili - protected -> public (renewable coral)
+         int i = random.nextInt(3) + 3;
+         int i1 = random.nextInt(3) + 3;
+         int i2 = random.nextInt(3) + 3;
