--- /dev/null
+++ b/net/minecraft/server/MinecraftServer.java.rej
@@ -1,0 +_,10 @@
+diff a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java	(rejected hunks)
+@@ -2033,6 +2033,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
+             runnable.run();
+         }
+ 
++        org.leavesmc.leaves.protocol.core.LeavesProtocolManager.handleTick(); // Leaves - protocol
++
+         profilerFiller.popPush("send chunks");
+ 
+         if (false) for (ServerPlayer serverPlayer : this.playerList.getPlayers()) { // Folia - region threading
