From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lophine <lophine@dev.null>
Date: Mon, 1 Jan 2026 00:00:00 +0800
Subject: [PATCH] Hopper idle optimization - field

Add idle tick counter field to HopperBlockEntity.

diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 0942fd82d4ea32f5c75f255ecd6c7b160851706e..0000000000000000000000000000000000000000 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -155,6 +155,7 @@
     public void setItem(int index, ItemStack stack) {
         this.unpackLootTable(null);
         this.getItems().set(index, stack);
         stack.limitSize(this.getMaxStackLeaves(stack)); // Leaves - item over-stack util
     }
+    int lophineHopperIdleTicks = 0; // Lophine - hopper idle optimization
 
     @Override
