From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 22 Mar 2026 01:39:17 +0800
Subject: [PATCH] Carpet features


diff --git a/src/main/java/me/earthme/luminol/config/ConfigManager.java b/src/main/java/me/earthme/luminol/config/ConfigManager.java
index afa7df571dc2adaf76020ef20ea794ebf2194692..842ee552b179e7cabefcfccf4775b4a2d393cd3d 100644
--- a/src/main/java/me/earthme/luminol/config/ConfigManager.java
+++ b/src/main/java/me/earthme/luminol/config/ConfigManager.java
@@ -25,6 +25,13 @@ public class ConfigManager {
     public static void initConfigs() {
         configfiles.put("luminol", ConfigsInstance.of("luminol", "me.earthme.luminol.config.modules"));
         configfiles.put("Mili", ConfigsInstance.of("Mili", "fun.bm.mili.config.modules")); // add Mili global config
+        configfiles.put("mili_carpet", ConfigsInstance.of(
+                new java.io.File("mili_config"),
+                "mili_carpet",
+                "mili_carpet_config.toml",
+                "milicarpetconfig",
+                "fun.bm.mili.carpet.config.modules"
+        )); // add mili carpet config
         preLoad();
     }
 

